{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da5aabfc-e467-46d4-a09a-6e8c89f5d684",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import *\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", message=\"Used element '.' for Atom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfb14584-e84d-42f7-8c60-077c7e380b6e",
   "metadata": {},
   "source": [
    "# Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f151586-70e4-447a-901c-71ba75943cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>(2, 40)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3r0hG01</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>2.60</td>\n",
       "      <td>3r0h</td>\n",
       "      <td>TAEIKPNKKILIELKVEKKPMGVIVCGGKNNHVTTGCVITHVYPEG...</td>\n",
       "      <td>[(479, 579)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1aqcA00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>2.30</td>\n",
       "      <td>1aqc</td>\n",
       "      <td>EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...</td>\n",
       "      <td>[(324, 488)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6hjfA02</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>310</td>\n",
       "      <td>10</td>\n",
       "      <td>1.70</td>\n",
       "      <td>6hjf</td>\n",
       "      <td>VPSFLYQQDVVVVLPKPYGEVRVDIAFGGNFFAIVPAEQLGIDISV...</td>\n",
       "      <td>[(189, 367)]</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1o1zA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>190</td>\n",
       "      <td>1.60</td>\n",
       "      <td>1o1z</td>\n",
       "      <td>HHHHVIVLGHRGYSAKYLENTLEAFMKAIEAGANGVELDVRLSKDG...</td>\n",
       "      <td>[(-3, 222)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cath_id  class  architecture  topology  superfamily  \\\n",
       "0  2w3sB01      3            90      1170           50   \n",
       "1  3be3A00      2            30        30          320   \n",
       "2  3zq4C03      3            10        20          580   \n",
       "3  1peqA03      1            10      1650           20   \n",
       "4  1bdoA00      2            40        50          100   \n",
       "5  3r0hG01      2            30        42           10   \n",
       "6  1aqcA00      2            30        29           30   \n",
       "7  2gnxA01      1            10      3450           30   \n",
       "8  6hjfA02      3            10       310           10   \n",
       "9  1o1zA00      3            20        20          190   \n",
       "\n",
       "   resolution_in_angstroms pdb_id  \\\n",
       "0                     2.60   2w3s   \n",
       "1                     2.04   3be3   \n",
       "2                     3.00   3zq4   \n",
       "3                     2.80   1peq   \n",
       "4                     1.80   1bdo   \n",
       "5                     2.60   3r0h   \n",
       "6                     2.30   1aqc   \n",
       "7                     2.45   2gnx   \n",
       "8                     1.70   6hjf   \n",
       "9                     1.60   1o1z   \n",
       "\n",
       "                                           sequences  \\\n",
       "0  SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1  QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2  DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "3  DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4  EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "5  TAEIKPNKKILIELKVEKKPMGVIVCGGKNNHVTTGCVITHVYPEG...   \n",
       "6  EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...   \n",
       "7                                                NaN   \n",
       "8  VPSFLYQQDVVVVLPKPYGEVRVDIAFGGNFFAIVPAEQLGIDISV...   \n",
       "9  HHHHVIVLGHRGYSAKYLENTLEAFMKAIEAGANGVELDVRLSKDG...   \n",
       "\n",
       "               cath_indices architecture_id  label  \n",
       "0                [(2, 124)]         (3, 90)      9  \n",
       "1                 [(6, 81)]         (2, 30)      2  \n",
       "2              [(449, 555)]         (3, 10)      5  \n",
       "3              [(294, 346)]         (1, 10)      0  \n",
       "4               [(77, 156)]         (2, 40)      3  \n",
       "5              [(479, 579)]         (2, 30)      2  \n",
       "6              [(324, 488)]         (2, 30)      2  \n",
       "7  [(101, 295), (431, 440)]         (1, 10)      0  \n",
       "8              [(189, 367)]         (3, 10)      5  \n",
       "9               [(-3, 222)]         (3, 20)      6  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('../data/cath_w_seqs_share.csv', index_col=0)\n",
    "data['cath_indices'] = data['cath_indices'].apply(safe_eval) # Convert `cath_indices` from str to list of tuples\n",
    "\n",
    "# Map architecture IDs to names and labels\n",
    "architecture_names = {\n",
    "    (1, 10): {\"name\": \"Mainly Alpha: Orthogonal Bundle\", \"label\": 0},\n",
    "    (1, 20): {\"name\": \"Mainly Alpha: Up-down Bundle\", \"label\": 1},\n",
    "    (2, 30): {\"name\": \"Mainly Beta: Roll\", \"label\": 2},\n",
    "    (2, 40): {\"name\": \"Mainly Beta: Beta Barrel\", \"label\": 3},\n",
    "    (2, 60): {\"name\": \"Mainly Beta: Sandwich\", \"label\": 4},\n",
    "    (3, 10): {\"name\": \"Alpha Beta: Roll\", \"label\": 5},\n",
    "    (3, 20): {\"name\": \"Alpha Beta: Alpha-Beta Barrel\", \"label\": 6},\n",
    "    (3, 30): {\"name\": \"Alpha Beta: 2-Layer Sandwich\", \"label\": 7},\n",
    "    (3, 40): {\"name\": \"Alpha Beta: 3-Layer(aba) Sandwich\", \"label\": 8},\n",
    "    (3, 90): {\"name\": \"Alpha Beta: Alpha-Beta Complex\", \"label\": 9}\n",
    "}\n",
    "\n",
    "# Create 'architecture_id' and 'label' columns directly from architecture_names\n",
    "data['architecture_id'] = data.apply(lambda row: (row['class'], row['architecture']), axis=1)\n",
    "data['label'] = data.apply(lambda row: architecture_names[(row['class'], row['architecture'])]['label'], axis=1)\n",
    "\n",
    "\n",
    "data.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4216addd-f01b-4236-925a-ca43aa4568a6",
   "metadata": {},
   "source": [
    "# File Hierarchy Construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df668978-5c31-4fb0-9dc5-a5a5f67ec9b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in data.iterrows():\n",
    "    cath_id = row['cath_id']\n",
    "    \n",
    "    # Create ../data/{cath_id}/pdb directory if it doesn't exist\n",
    "    pdb_dir = f\"../data/{cath_id}/pdb\"\n",
    "    os.makedirs(pdb_dir, exist_ok=True)\n",
    "    \n",
    "    # Create ../data/{cath_id}/networkx directory if it doesn't exist\n",
    "    networkx_dir = f\"../data/{cath_id}/networkx\"\n",
    "    os.makedirs(networkx_dir, exist_ok=True)\n",
    "\n",
    "    # Create ../data/{cath_id}/networkx directory if it doesn't exist\n",
    "    seq_dir = f\"../data/{cath_id}/seqs\"\n",
    "    os.makedirs(seq_dir, exist_ok=True)\n",
    "    \n",
    "    # Copy file from ../data/pdb_share/{cath_id} to ../data/{cath_id}/pdb\n",
    "    source_file = f\"../data/pdb_share/{cath_id}\"\n",
    "    dest_file = f\"../data/{cath_id}/pdb/{cath_id}\"\n",
    "    shutil.copyfile(source_file, dest_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e18fe9d5-7c54-4d0d-b20f-bc594386fe88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def safe_len(x):\n",
    "#     if isinstance(x, (list, tuple)):\n",
    "#         return len(x)\n",
    "#     return 0  # or you might want to return a different value for non-list/tuple entries\n",
    "\n",
    "# subset_df = data[data['cath_indices'].apply(safe_len) > 1]\n",
    "# subset_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "35ab1e53-e681-48b5-ad7f-0f6ce75b3979",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # PDB: 4adi\n",
    "# # cath_id: 4adiA01\n",
    "# # iloc: 33\n",
    "# # cath_indices: [(2, 30), (164, 198), (319, 331)]\n",
    "# uniprot_seq = 'MASTTPITMEDLQKALEAQSRALRAGLAAGASQSRRPRPPRQRDSSTSGDDSGRDSGGPRRRRGNRGRGQRKDWSRAPPPPEERQESRSQTPAPKPSRAPPQQPQPPRMQTGRGGSAPRPELGPPTNPFQAAVARGLRPPLHDPDTEAPTEACVTSWLWSEGEGAVFYRVDLHFTNLGTPPLDEDGRWDPALMYNPCGPEPPAHVVRAYNQPAGDVRGVWGKGERTYAEQDFRVGGTRWHRLLRMPVRGLDGDTAPLPPHTTERIETRSARHPWRIRFGAPQAFLAGLLLAAVAVGTARAGLQPRADMAAPPMPPQPPRAHGQHYGHHHHQLPFLGHDGHHGGTLRVGQHHRNASDVLPGHWLQGGWGCYNLSDWHQGTHVCHTKHMDFWCVEHDRPPPATPTSLTTAANSTTAATPATAPPPCHAGLNDSCGGFLSGCGPMRLRHGADTRCGRLICGLSTTAQYPPTRFGCAMRWGLPPWELVVLTARPEDGWTCRGVPAHPGTRCPELVSPMGRATCSPASALWLATANALSLDHAFAAFVLLVPWVLIFMVCRRACRRRGAAAALTAVVLQGYNPPAYGEEAFTYLCTAPGCATQTPVPVRLAGVRFESKIVDGGCFAPWDLEATGACICEIPTDVSCEGLGAWVPTAPCARIWNGTQRACTFWAVNAYSSGGYAQLASYFNPGGSYYKQYHPTACEVEPAFGHSDAACWGFPTDTVMSVFALASYVQHPHKTVRVKFHTETRTVWQLSVAGVSCNVTTEHPFCNTPHGQLEVQVPPDPGDLVEYIMNYTGNQQSRWGLGSPNCHGPDWASPVCQRHSPDCSRLVGATPERPRLRLVDADDPLLRTAPGPGEVWVTPVIGSQARKCGLHIRAGPYGHATVEMPEWIHAHTTSDPWHPPGPLGLKFKTVRPVALPRALAPPRNVRVTGCYQCGTPALVEGLAPGGGNCHLTVNGEDVGAFPPGKFVTAALLNTPPPYQVSCGGESDRASARVIDPAAQSFTGVVYGTHTTAVSETRQTWAEWAAAHWWQLTLGAICALLLAGLLACCAKCLYYLRGAIAPR'\n",
    "# print(uniprot_seq[1:29])\n",
    "# print(uniprot_seq[163:197])\n",
    "# print(uniprot_seq[318:330])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "91bcc69a-576b-4483-b851-2c2d27366cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data.iloc[33]['sequences']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dbf63256-4e95-4584-836f-f3daf1ec6f37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # PDB: 2gnx\n",
    "# # cath_id: 2gnxA01\n",
    "# # iloc: 7\n",
    "# # cath_indices: [(101, 295), (431, 440)]\n",
    "# seq = 'MGESIPLAAPVPVEQAVLETFFSHLGIFSYDKAKDNVEKEREANKSAGGSWLSLLAALAHLAAAEKVYHSLTYLGQKLGGQSFFSRKDSIRTIYTSLHNELKKVVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKMYALSTQKFINTEELVSTLDTILRKYSSRFHHPILSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRETKKHLFGGQSQKAVQPPHLFLWLMKLKTMLLAKFSFYFHEALSRQTTASEMKALTAKANPDLFGKISSFIRKYDAANVSLIFDNRGSESFQGHGYHHPHSYREAPKGVDQYPAVVSLPSDRPVMHWPNVIMIMTDRASDLNSLEKVVHFYDDKVQSTYFLTRPEPHFTIVVIFESKKSERDSHFISFLNELSLALKNPKVFASLKPGSKG'\n",
    "# a = seq[100:294]\n",
    "# b = seq[430:439]\n",
    "# print(a + b)\n",
    "# print(len(a+b))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5032597e-9c86-4448-a15f-88bf75e644f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cath_indices = data.iloc[7]['cath_indices']\n",
    "# print(cath_indices)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12348b1a-2104-493d-b21b-304e44b28ba8",
   "metadata": {},
   "source": [
    "# Handling Missing Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "343068da-533c-48a8-a85d-4d773423d736",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>19hcA01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.80</td>\n",
       "      <td>19hc</td>\n",
       "      <td>CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>2fd4A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>1.80</td>\n",
       "      <td>2fd4</td>\n",
       "      <td>KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>2xskA00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2420</td>\n",
       "      <td>1.70</td>\n",
       "      <td>2xsk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(3, 97)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3138</th>\n",
       "      <td>5inwA02</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>10</td>\n",
       "      <td>2.70</td>\n",
       "      <td>5inw</td>\n",
       "      <td>WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3838</th>\n",
       "      <td>1dkiC01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>70</td>\n",
       "      <td>50</td>\n",
       "      <td>1.60</td>\n",
       "      <td>1dki</td>\n",
       "      <td>FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4330</th>\n",
       "      <td>1kskA01</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>290</td>\n",
       "      <td>10</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1ksk</td>\n",
       "      <td>GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4668</th>\n",
       "      <td>4dgwC00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2690</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4dgw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 253)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>1nthA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>460</td>\n",
       "      <td>1.55</td>\n",
       "      <td>1nth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>3pieA02</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>1370</td>\n",
       "      <td>250</td>\n",
       "      <td>2.90</td>\n",
       "      <td>3pie</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(379, 487)]</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "7     2gnxA01      1            10      3450           30   \n",
       "178   19hcA01      3            90        10           10   \n",
       "2208  2fd4A00      3            30        40          110   \n",
       "2827  2xskA00      2            60        40         2420   \n",
       "3138  5inwA02      2            30        39           10   \n",
       "3838  1dkiC01      3            90        70           50   \n",
       "4330  1kskA01      3            10       290           10   \n",
       "4668  4dgwC00      2            60        40         2690   \n",
       "5776  1nthA00      3            20        20          460   \n",
       "6010  3pieA02      3            30      1370          250   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "7                        2.45   2gnx   \n",
       "178                      1.80   19hc   \n",
       "2208                     1.80   2fd4   \n",
       "2827                     1.70   2xsk   \n",
       "3138                     2.70   5inw   \n",
       "3838                     1.60   1dki   \n",
       "4330                     2.00   1ksk   \n",
       "4668                     3.11   4dgw   \n",
       "5776                     1.55   1nth   \n",
       "6010                     2.90   3pie   \n",
       "\n",
       "                                              sequences  \\\n",
       "7                                                   NaN   \n",
       "178   CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...   \n",
       "2208  KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...   \n",
       "2827                                                NaN   \n",
       "3138  WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...   \n",
       "3838  FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...   \n",
       "4330  GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...   \n",
       "4668                                                NaN   \n",
       "5776                                                NaN   \n",
       "6010                                                NaN   \n",
       "\n",
       "                  cath_indices architecture_id  label  \n",
       "7     [(101, 295), (431, 440)]         (1, 10)      0  \n",
       "178                        NaN         (3, 90)      9  \n",
       "2208                       NaN         (3, 30)      7  \n",
       "2827                 [(3, 97)]         (2, 60)      4  \n",
       "3138                       NaN         (2, 30)      2  \n",
       "3838                       NaN         (3, 90)      9  \n",
       "4330                       NaN         (3, 10)      5  \n",
       "4668              [(101, 253)]         (2, 60)      4  \n",
       "5776                [(2, 458)]         (3, 20)      6  \n",
       "6010              [(379, 487)]         (3, 30)      7  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify rows with any NaN values\n",
    "rows_with_nan = data[data.isna().any(axis=1)]\n",
    "rows_with_nan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57a1986f-45b4-453e-bdad-412a9a474ece",
   "metadata": {},
   "source": [
    "### CASE 1: Rows that have `sequences` but NaN `cath_indices` can be automatically corrected with residue indices present in the provided pdb files using a call to helper function `extract_resid_ranges()` which considers contiguous ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "443e839e-9a51-4648-858b-5a81def74ad9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>19hcA01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.8</td>\n",
       "      <td>19hc</td>\n",
       "      <td>CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>2fd4A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>1.8</td>\n",
       "      <td>2fd4</td>\n",
       "      <td>KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3138</th>\n",
       "      <td>5inwA02</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>10</td>\n",
       "      <td>2.7</td>\n",
       "      <td>5inw</td>\n",
       "      <td>WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3838</th>\n",
       "      <td>1dkiC01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>70</td>\n",
       "      <td>50</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1dki</td>\n",
       "      <td>FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4330</th>\n",
       "      <td>1kskA01</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>290</td>\n",
       "      <td>10</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1ksk</td>\n",
       "      <td>GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "178   19hcA01      3            90        10           10   \n",
       "2208  2fd4A00      3            30        40          110   \n",
       "3138  5inwA02      2            30        39           10   \n",
       "3838  1dkiC01      3            90        70           50   \n",
       "4330  1kskA01      3            10       290           10   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "178                       1.8   19hc   \n",
       "2208                      1.8   2fd4   \n",
       "3138                      2.7   5inw   \n",
       "3838                      1.6   1dki   \n",
       "4330                      2.0   1ksk   \n",
       "\n",
       "                                              sequences cath_indices  \\\n",
       "178   CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...          NaN   \n",
       "2208  KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...          NaN   \n",
       "3138  WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...          NaN   \n",
       "3838  FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...          NaN   \n",
       "4330  GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...          NaN   \n",
       "\n",
       "     architecture_id  label  \n",
       "178          (3, 90)      9  \n",
       "2208         (3, 30)      7  \n",
       "3138         (2, 30)      2  \n",
       "3838         (3, 90)      9  \n",
       "4330         (3, 10)      5  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case1 = data[data['sequences'].notna() & data['cath_indices'].isna()]\n",
    "case1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e60ddd3b-8843-45ba-aa07-62b2e4a9eb3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated `cath_indices` for cath_id 19hcA01:\n",
      "[(100, 129), (168, 292)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 2fd4A00:\n",
      "[(435, 553)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 5inwA02:\n",
      "[(251, 348), (408, 416)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 1dkiC01:\n",
      "[(1, 112), (154, 253)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 1kskA01:\n",
      "[(1, 63)]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Define the threshold value for grouping contiguous residue ranges.\n",
    "Residue ranges are considered contiguous if the difference between consecutive residue IDs\n",
    "does not exceed this threshold value.\n",
    "This threshold value directly influences the identification of gaps in the next section, so it is important \n",
    "to choose this number carefully.\n",
    "'''\n",
    "gap_threshold = 20\n",
    "\n",
    "# Iterate over case1, apply get_sequence_from_pdb, and update data\n",
    "for index, row in case1.iterrows():\n",
    "    # Data accession\n",
    "    cath_id = row['cath_id']\n",
    "    pdb_filename = f\"../data/pdb_share/{cath_id}\"\n",
    "\n",
    "    # Call provided helper fxn\n",
    "    cath_indices = extract_resid_ranges(pdb_filename, threshold=gap_threshold)\n",
    "\n",
    "    # Update the data frame to remove case2 NaN values\n",
    "    data.at[index, 'cath_indices'] = cath_indices\n",
    "    print(f\"Updated `cath_indices` for cath_id {cath_id}:\\n{cath_indices}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e405b4c6-1830-45fe-86e8-02b1fdcf904b",
   "metadata": {},
   "source": [
    "### CASE 2: Rows that have `cath_indices` but NaN `sequences` can be automatically corrected with the sequence present in UniProtKB\n",
    "- Note: Helper function `get_sequence_from_pdb()` does not suffice here because of the prescence of `UNK` resnames in case2 pdb files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "922472af-3e33-4263-ad08-20843e316cb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>2xskA00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2420</td>\n",
       "      <td>1.70</td>\n",
       "      <td>2xsk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(3, 97)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4668</th>\n",
       "      <td>4dgwC00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2690</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4dgw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 253)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>1nthA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>460</td>\n",
       "      <td>1.55</td>\n",
       "      <td>1nth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>3pieA02</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>1370</td>\n",
       "      <td>250</td>\n",
       "      <td>2.90</td>\n",
       "      <td>3pie</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(379, 487)]</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "7     2gnxA01      1            10      3450           30   \n",
       "2827  2xskA00      2            60        40         2420   \n",
       "4668  4dgwC00      2            60        40         2690   \n",
       "5776  1nthA00      3            20        20          460   \n",
       "6010  3pieA02      3            30      1370          250   \n",
       "\n",
       "      resolution_in_angstroms pdb_id sequences              cath_indices  \\\n",
       "7                        2.45   2gnx       NaN  [(101, 295), (431, 440)]   \n",
       "2827                     1.70   2xsk       NaN                 [(3, 97)]   \n",
       "4668                     3.11   4dgw       NaN              [(101, 253)]   \n",
       "5776                     1.55   1nth       NaN                [(2, 458)]   \n",
       "6010                     2.90   3pie       NaN              [(379, 487)]   \n",
       "\n",
       "     architecture_id  label  \n",
       "7            (1, 10)      0  \n",
       "2827         (2, 60)      4  \n",
       "4668         (2, 60)      4  \n",
       "5776         (3, 20)      6  \n",
       "6010         (3, 30)      7  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case2 = data[data['sequences'].isna() & data['cath_indices'].notna()]\n",
    "case2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2cdf73ae-4857-4292-8e8d-1753f5ef6791",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cath_indices = [(101, 295), (431, 440)]\n",
    "\n",
    "# for idx_range in cath_indices:\n",
    "#     start_idx, end_idx = idx_range\n",
    "#     idx_range_set = set(range(start_idx, end_idx + 1))  # Create a set of the range for quick lookup\n",
    "#     print(idx_range_set)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "65becb91-4c84-4f29-adf5-32c248223070",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from Bio.Align import PairwiseAligner\n",
    "\n",
    "# def extract_aligned_segment(given_sequence, ground_truth_sequence):\n",
    "#     \"\"\"\n",
    "#     Extracts the segment from the ground truth sequence that most aligns with the given sequence\n",
    "#     and replaces the 'U' amino acids with the corresponding amino acids from the ground truth sequence.\n",
    "    \n",
    "#     Args:\n",
    "#     given_sequence (str): The sequence to be aligned.\n",
    "#     ground_truth_sequence (str): The reference sequence to align against.\n",
    "    \n",
    "#     Returns:\n",
    "#     str: The given sequence with 'U' amino acids replaced by aligned ground truth amino acids.\n",
    "#     \"\"\"\n",
    "    \n",
    "#     # Initialize the aligner\n",
    "#     aligner = PairwiseAligner()\n",
    "#     aligner.mode = 'global'\n",
    "    \n",
    "#     # Perform the alignment\n",
    "#     alignments = aligner.align(given_sequence.replace('U', ''), ground_truth_sequence)\n",
    "    \n",
    "#     # Get the best alignment\n",
    "#     best_alignment = alignments[0]\n",
    "    \n",
    "#     # Extract the aligned sequences\n",
    "#     aligned_given = best_alignment.aligned[0]\n",
    "#     aligned_truth = best_alignment.aligned[1]\n",
    "    \n",
    "#     # Construct the aligned segment from the ground truth sequence\n",
    "#     aligned_segment = \"\"\n",
    "#     for i in range(len(aligned_given)):\n",
    "#         start_given, end_given = aligned_given[i]\n",
    "#         start_truth, end_truth = aligned_truth[i]\n",
    "        \n",
    "#         if given_sequence[start_given] != 'U':\n",
    "#             aligned_segment += ground_truth_sequence[start_truth:end_truth]\n",
    "#         else:\n",
    "#             aligned_segment += given_sequence[start_given:end_given]\n",
    "\n",
    "#     # Print the aligned segment\n",
    "#     print(f\"Aligned segment from the ground truth sequence: {aligned_segment}\")\n",
    "    \n",
    "#     # Print the alignment for reference\n",
    "#     print(best_alignment)\n",
    "\n",
    "#     # Print detailed alignment positions and segments\n",
    "#     print(\"Aligned positions in given sequence:\")\n",
    "#     for start, end in aligned_given:\n",
    "#         print(f\"Start: {start}, End: {end}, Segment: {given_sequence[start:end]}\")\n",
    "    \n",
    "#     print(\"Aligned positions in ground truth sequence:\")\n",
    "#     for start, end in aligned_truth:\n",
    "#         print(f\"Start: {start}, End: {end}, Segment: {ground_truth_sequence[start:end]}\")\n",
    "\n",
    "#     return aligned_segment\n",
    "\n",
    "# # Example usage\n",
    "# ground_truth_sequence = \"PHLSEQLCFFVQAREIADFYEKYALSTQKFINTEELVSTLDTILRKYSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRPPHLFLWLKLKTLLAKFSFYFHEALSRQTTASEKALTAKANP\"\n",
    "#                         \"PHLSEQLCFFVQAREIADFYEKYALSTQKFINTEELVSTLDTILRKYSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRPPHLFLWLKLKTLLAKFSFYFHEALSRQTTASEKALTAKANP\"\n",
    "# given_sequence = \"MGESIPLAAPVPVEQAVLETFFSHLGIFSYDKAKDNVEKEREANKSAGGSWLSLLAALAHLAAAEKVYHSLTYLGQKLGGQSFFSRKDSIRTIYTSLHNELKKVVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKMYALSTQKFINTEELVSTLDTILRKYSSRFHHPILSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRETKKHLFGGQSQKAVQPPHLFLWLMKLKTMLLAKFSFYFHEALSRQTTASEMKALTAKANPDLFGKISSFIRKYDAANVSLIFDNRGSESFQGHGYHHPHSYREAPKGVDQYPAVVSLPSDRPVMHWPNVIMIMTDRASDLNSLEKVVHFYDDKVQSTYFLTRPEPHFTIVVIFESKKSERDSHFISFLNELSLALKNPKVFASLKPGSKG\"\n",
    "\n",
    "# aligned_segment = extract_aligned_segment(given_sequence, ground_truth_sequence)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "10a922d6-9aba-4f4d-8817-e33283469e22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CATH_ID] 2gnxA01\n",
      "\n",
      "Subject:  (UniProt accession Q6P1I3)\n",
      "MGESIPLAAPVPVEQAVLETFFSHLGIFSYDKAKDNVEKEREANKSAGGSWLSLLAALAHLAAAEKVYHSLTYLGQKLGGQSFFSRKDSIRTIYTSLHNELKKVVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKMYALSTQKFINTEELVSTLDTILRKYSSRFHHPILSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRETKKHLFGGQSQKAVQPPHLFLWLMKLKTMLLAKFSFYFHEALSRQTTASEMKALTAKANPDLFGKISSFIRKYDAANVSLIFDNRGSESFQGHGYHHPHSYREAPKGVDQYPAVVSLPSDRPVMHWPNVIMIMTDRASDLNSLEKVVHFYDDKVQSTYFLTRPEPHFTIVVIFESKKSERDSHFISFLNELSLALKNPKVFASLKPGSKG\n",
      "\n",
      "Given Seq:\n",
      "Case 2 - nan\n",
      "Building sequence from pdb_share...\n",
      "\n",
      "[GAP DETECTED for cath_id 2gnxA01]\n",
      "Missing residues in range 101 - 295: [122, 137, 146, 173, 174, 175, 176, 177, 178, 179, 180, 181, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 259, 264, 286]\n",
      "\n",
      "Repairing...\n",
      "\n",
      "Query Sequence 1:  (101, 295)\n",
      "UUUUUUUUUUUUUUUUUUUUUUPHLSEQLCFFVQARUEIADFYEKUYALSTQKFINTEELVSTLDTILRKYSUUUUUUUUUPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRUUUUUUUUUUUUUUUUPPHLFLWLUKLKTULLAKFSFYFHEALSRQTTASEUKALTAKANP\n",
      "\n",
      "Alignment Segment (from subject):\n",
      "VVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKMYALSTQKFINTEELVSTLDTILRKYSSRFHHPILSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRETKKHLFGGQSQKAVQPPHLFLWLMKLKTMLLAKFSFYFHEALSRQTTASEMKALTAKANP\n",
      "\n",
      "WARNING: Length of aligned segment [192] does not match the expected number of amino acids [195]\n",
      "\n",
      "Trying a right shift of the aligned segment.\n",
      "Right shift solve was successful.\n",
      "\n",
      "A match between the shifted `aligned_segment` and the `query_seq` was found:\n",
      "UUUUUUUUUUUUUUUUUUUUUUPHLSEQLCFFVQARUEIADFYEKUYALSTQKFINTEELVSTLDTILRKYSUUUUUUUUUPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRUUUUUUUUUUUUUUUUPPHLFLWLUKLKTULLAKFSFYFHEALSRQTTASEUKALTAKANP\n",
      "+++VVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKMYALSTQKFINTEELVSTLDTILRKYSSRFHHPILSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRETKKHLFGGQSQKAVQPPHLFLWLMKLKTMLLAKFSFYFHEALSRQTTASEMKALTAKANP\n",
      "\n",
      "Replacing '+' values in the shifted `aligned_segment` with amino acids codes from the `subject_seq`.\n",
      "\n",
      "Successfully repaired the aligned segment with the manual solve.\n",
      "Repaired aligned segment length: 195\n",
      "\n",
      "\n",
      "Query Sequence 2:  (431, 440)\n",
      "KNPKVFASLK\n",
      "\n",
      "Alignment Segment (from subject):\n",
      "KNPKVFASLK\n",
      "\n",
      "[REPAIRED]\n",
      "Updated FASTA sequence for CATH ID 2gnxA01 :\n",
      "LKKVVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKMYALSTQKFINTEELVSTLDTILRKYSSRFHHPILSPLESSFQLEVGVLSHLLKAQAQISEWKFLPSLVTLHNAHTKLQSWGQTFEKQRETKKHLFGGQSQKAVQPPHLFLWLMKLKTMLLAKFSFYFHEALSRQTTASEMKALTAKANPKNPKVFASLK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result = []\n",
    "    \n",
    "    # Data accession\n",
    "    pdb_code = '2gnx'\n",
    "    cath_id = '2gnxA01'\n",
    "    cath_indices = [(101, 295), (431, 440)]\n",
    "    given_seq = data[data['cath_id'] == '2gnxA01']['sequences'].item()\n",
    "    print(f'[CATH_ID] {cath_id}\\n')\n",
    "\n",
    "    # Web scraping fxn calls\n",
    "    uniprot_accession = get_uniprot_accession_from_pdb(pdb_code)\n",
    "    subject_sequence = get_fasta_sequence_from_uniprot(uniprot_accession)\n",
    "    print(f'Subject:  (UniProt accession {uniprot_accession})\\n{subject_sequence}\\n')\n",
    "    subject_file = f'../data/{cath_id}/seqs/subject.fasta'\n",
    "    with open(subject_file, 'w') as t:\n",
    "        t.write(subject_sequence)\n",
    "\n",
    "    if pd.isna(given_seq):    \n",
    "        print(f'Given Seq:\\nCase 2 - {given_seq}\\nBuilding sequence from pdb_share...\\n')\n",
    "    else:\n",
    "        print(f'Given Seq:\\n{given_seq}\\n')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    pdb_filename = f\"../data/pdb_share/{cath_id}\"\n",
    "    for i, idx_range in enumerate(cath_indices):\n",
    "        # The query sequence corresponds to the protein domain specific to the idx_range tuple\n",
    "        query_sequence = get_domain_sequence_from_pdb(pdb_filename, idx_range, cath_id) \n",
    "        print(f'\\nQuery Sequence {i+1}:  {idx_range}\\n{query_sequence}\\n')\n",
    "        query_file = f'../data/{cath_id}/seqs/query_{i+1}.fasta'\n",
    "        with open(query_file, 'w') as q:\n",
    "            q.write(query_sequence)\n",
    "\n",
    "        # Perform sequence alignment for this domain segment using NCBI Blast+\n",
    "        output_file = f'../data/{cath_id}/seqs/output_{i+1}.txt'\n",
    "        aligned_segment = extract_aligned_segment(query_sequence, subject_sequence, query_file, subject_file, output_file)\n",
    "        print(f'Alignment Segment (from subject):\\n{aligned_segment}\\n')\n",
    "\n",
    "        # Error Handling - Sequence alignment w/ BLAST+ sometimes leaves out amino acids. If not handled, this will corrupt remodeling.\n",
    "        expected_num_AAs = idx_range[1] - idx_range[0] + 1\n",
    "        if len(aligned_segment) != expected_num_AAs:\n",
    "            print(f\"WARNING: Length of aligned segment [{len(aligned_segment)}] does not match the expected number of amino acids [{expected_num_AAs}]\\n\")\n",
    "            # Manual sequence alignment\n",
    "            aligned_segment = manual_alignment(query_sequence, aligned_segment, subject_sequence)\n",
    "\n",
    "        result.append(aligned_segment)\n",
    "\n",
    "\n",
    "    # Concatenate the indexed fasta sequences if len(cath_indices) > 1\n",
    "    updated_sequence = ''.join(result)\n",
    "\n",
    "    # Update the data frame to remove case2 NaN values\n",
    "    data.at[index, 'sequences'] = updated_sequence\n",
    "    print('[REPAIRED]')\n",
    "    print(f\"Updated FASTA sequence for CATH ID {cath_id} :\\n{updated_sequence}\\n\")\n",
    "        \n",
    "except Exception as e:\n",
    "    print(f\"Error: {str(e)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8785145e-7726-45fa-9075-a67a93b4935e",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Iterate over case2, apply get_fasta_sequence_from_uniprot, and update data\n",
    "\n",
    "The indexing approach maps the provided `cath_indices` onto the UniProtKB sequence to extract the relevant segment\n",
    "'''\n",
    "for index, row in case2.iterrows():\n",
    "    try:\n",
    "        # Data accession\n",
    "        pdb_code = row['pdb_id']\n",
    "        # cath_indices = row['cath_indices']\n",
    "        cath_id = row['cath_id']\n",
    "\n",
    "\n",
    "        # Web scraping fxn calls\n",
    "        uniprot_accession = get_uniprot_accession_from_pdb(pdb_code)\n",
    "        ground_truth_fasta_sequence = get_fasta_sequence_from_uniprot(uniprot_accession)\n",
    "\n",
    "        pdb_filename = f\"../data/pdb_share/{cath_id}\"\n",
    "        available_fasta_sequence = get_sequence_from_pdb(pdb_filename)\n",
    "\n",
    "        aligned_segment = extract_aligned_segment(available_fasta_sequence, ground_truth_fasta_sequence)\n",
    "        print(aligned_segment)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "        \n",
    "       \n",
    "    except Exception as e:\n",
    "        print(f\"Failed to fetch sequence for PDB code {pdb_code}: {str(e)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c13f123-c264-4c0a-bbb3-3a5c6592530d",
   "metadata": {},
   "source": [
    "# Handling Sequence Gaps\n",
    "- Phase 1: Detect Gaps\n",
    "- Phase 2: Replace Gaps in `sequences` col w/ AAs known by UniProtKB\n",
    "- Phase 3: Protein domain remodeling w/ SCWRL4\n",
    "  #### Note: Case 2 data entries already have gaps corrected by the previous UniProtKB look up and do not need Phase 2 processing. They still require remodeling, however."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83c5599c-998f-4119-be0f-c097b5c24f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example_gaps = {'3zq4C03': [('C', 493, 501)],\n",
    "#                 '3i9v600': [('6', 58, 74)]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "366777b9-f374-43a2-80c9-7f0c0659d8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add a column to track whether or not a given sequence entry has a gap\n",
    "for index, row in data.iterrows():\n",
    "    cath_indices = row['cath_indices'] # grab the indices\n",
    "    cath_id = row['cath_id']\n",
    "    \n",
    "    # Open the PDB file\n",
    "    pdb_filename = f\"../data/pdb_share/{cath_id}\"\n",
    "    gap = detect_seq_gaps(pdb_filename, cath_indices, cath_id)\n",
    "\n",
    "    data.at[index, 'gap'] = int(gap) # Will be 1 or 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d77116-dc97-4886-b4e4-e0756de450e5",
   "metadata": {},
   "source": [
    "### Case study 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14b706e2-44ba-496a-9810-58f54bc81fcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "given_seq = data[data['cath_id'] == '3be3A00']['sequences'].item()\n",
    "print(given_seq)\n",
    "print(len(given_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68f5298a-3436-4be9-9352-742b3a6d22f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This accession originally has both `cath_indices` and `sequences`\n",
    "\n",
    "# PDB: 3be3\n",
    "# cath_id: 3be3A00\n",
    "# iloc: 1\n",
    "# cath_indices: [(6, 81)]\t\n",
    "# sequences (given): QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPSTRLLRIWNETVDTGAGPQPRFAYVGHVTPE\n",
    "\n",
    "# It is missing resid 50, and thus has a gap\n",
    "# [GAP DETECTED for cath_id 3be3A00] Missing residues in range 6 - 81: [50]\n",
    "\n",
    "uniprot_seq = 'MAMQDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPMSTRLLRIWNETVDTGAGPQPRFAYVGHVTPEQG'\n",
    "print(uniprot_seq[5:81])\n",
    "print(len(uniprot_seq))\n",
    "\n",
    "actual = uniprot_seq[3:79]\n",
    "print(actual)\n",
    "print(len(actual))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af13a9f3-a511-4e9a-ad91-a1144b19835f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# provided (Missing residue 50)\n",
    "# QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPSTRLLRIWNETVDTGAGPQPRFAYVGHVTPE\n",
    "\n",
    "# UniProtKB GT\n",
    "# MAMQDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPMSTRLLRIWNETVDTGAGPQPRFAYVGHVTPEQG\n",
    "\n",
    "# extracted\n",
    "# QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPMSTRLLRIWNETVDTGAGPQPRFAYVGHVTPE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff863f3b-7cf5-41c6-8b2b-2e0b2391de19",
   "metadata": {},
   "source": [
    "### Case study 2 - Indexing Worked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e47b95d4-07f3-418c-b15e-66611eaa0096",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[data['cath_id'] == '1vkwA02']['sequences'].item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc8efb4f-51c0-4430-8050-5c8157880ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This accession originally has both `cath_indices` and `sequences`\n",
    "\n",
    "# PDB: 1vkw\n",
    "# cath_id: 1vkwA02\n",
    "# iloc: 6222\n",
    "# cath_indices: [(111, 206)]\t\n",
    "# sequences (given): 'FTRKRRPITSFLENDLEELPPEIVKIVETILAPSALNRQPWKIKYTGGELCISSERPVDLGIALSHAYLTAREIFKREPVIQKRGEDTYCLILNP'\n",
    "\n",
    "# It is missing resid 139, and thus has a gap\n",
    "# [GAP DETECTED for cath_id 1vkwA02] Missing residues in range 111 - 206: [139]\n",
    "\n",
    "uniprot_seq = 'MNIFEAIENRHSVRDFLERKMPERVKDDIENLLVKFITKKLDWKINLSSFPSYIYAKAEKHFDELVEYGFQGEQIVLFLTAQGFGTCWMARSPHPDVPYIIVFGYPRTRNFTRKRRPITSFLENDLEELPPEIVKIVEMTILAPSALNRQPWKIKYTGGELCISSERPVDLGIALSHAYLTAREIFKREPVIQKRGEDTYCLILNP'\n",
    "corrected = uniprot_seq[110:206]\n",
    "print(corrected)\n",
    "print(len(corrected))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "456f009b-c55b-4963-8813-ca3b6b2b9b80",
   "metadata": {},
   "outputs": [],
   "source": [
    "line = 'FTRKRRPITSFLENDLEELPPEIVKIVETILAPSALNRQPWKIKYTGGELCISSERPVDLGIALSHAYLTAREIFKREPVIQKRGEDTYCLILNP'\n",
    "print(len(line))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9ceba08-cecd-4383-8716-9a426a2ff721",
   "metadata": {},
   "source": [
    "### Case study 3 - Indexing Worked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eed2820-1955-4750-895b-7a01a7b4bd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "given_seq = data[data['cath_id'] == '3ltiA01']['sequences'].item()\n",
    "print(given_seq)\n",
    "print(len(given_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b21f7a8-e682-4c7f-835b-20f9b1c2457c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This accession originally has both `cath_indices` and `sequences`\n",
    "\n",
    "# PDB: 3lti\n",
    "# cath_id: 3ltiA01\n",
    "# iloc: 6127\n",
    "# cath_indices: [(152, 224), (343, 443)]\n",
    "# sequences (given): \n",
    "# SPGVFFDSDVLYNARIIPYRGSWLDFEFDPKDNLFVRIDRRRKLPATIILRALNYTTEQILDLFHGPYISETLRVDPTNDRLSALVEIYRRPGEPPTREAAESLFENLFFSEDRYDLSAVGRFNRSLLREEIEGSGILSDDIIDVLIDIRNGGEVD\n",
    "\n",
    "# It is missing many residues, and thus has a gap\n",
    "\n",
    "# [GAP DETECTED for cath_id 3ltiA01] Missing residues in range 152 - 224: [161, 162, 163, 164, 165, 166, 167, 168, 169]\n",
    "# [GAP DETECTED for cath_id 3ltiA01] Missing residues in range 343 - 443: [369, 370, 403, 404, 422, 429, 430, 431, 439]\n",
    "\n",
    "uniprot_seq = 'MVYSYTEKKRIRKDFGKRPQVLDVPYLLSIQLDSFQKFIEQDPEGQYGLEAAFRSVFPIQSYSGNSELQYVSYRLGEPVFDVQECQIRGVTYSAPLRVKLRLVIYEREAPEGTVKDIKEQEVYMGEIPLMTDNGTFVINGTERVIVSQLHRSPGVFFDSDKGKTHSSGKVLYNARIIPYRGSWLDFEFDPKDNLFVRIDRRRKLPATIILRALNYTTEQILDLFFEKVIFEIRDNKLQMELVPERLRGETASFDIEANGKVYVEKGRRITARHIRQLEKDDVKLIEVPVEYIAGKVVAKDYIDESTGELICAANMELSLDLLAKLSQSGHKRIETLFTNDLDHGPYISETLRVDPTNDRLSALVEIYRMMRPGEPPTREAAESLFENLFFSEDRYDLSAVGRMKFNRSLLREEIEGSGILSKDDIIDVMKKLIDIRNGKGEVDDIDHLGNRRIRSVGEMAENQFRVGLVRVERAVKERLSLGDLDTLMPQDMINAKPISAAVKEFFGSSQLSQFMDQNNPLSEITHKRRISALGPGGLTRERAGFEVRDVHPTHYGRVCPIETPEGPNIGLINSLSVYAQTNEYGFLETPYRKVTDGVVTDEIHYLSAIEEGNYVIAQANSNLDEEGHFVEDLVTCRSKGESSLFSRDQVDYMDVSTQQVVSVGASLIPFLEHDDANRALMGANMQRQAVPTLRADKPLVGTGMERAVAVDSGVTAVAKRGGVVQYVDASRIVIKVNEDEMYPGEAGIDIYNLTKYTRSNQNTCINQMPCVSLGEPVERGDVLADGPSTDLGELALGQNMRVAFMPWNGYNFEDSILVSERVVQEDRFTTIHIQELACVSRDTKLGPEEITADIPNVGEAALSKLDESGIVYIGAEVTGGDILVGKVTPKGETQLTPEEKLLRAIFGEKASDVKDSSLRVPNGVSGTVIDVQVFTRDGVEKDKRALEIEEMQLKQAKKDLSEELQILEAGLFSRIRAVLVAGGVEAEKLDKLPRDRWLELGLTDEEKQNQLEQLAEQYDELKHEFEKKLEAKRRKITQGDDLAPGVLKIVKVYLAVKRRIQPGDKMAGRHGNKGVISKINPIEDMPYDENGTPVDIVLNPLGVPSRMNIGQILETHLGMAAKGIGDKINAMLKQQQEVAKLREFIQRAYDLGADVRQKVDLSTFSDEEVMRLAENLRKGMPIATPVFDGAKEAEIKELLKLGDLPTSGQIRLYDGRTGEQFERPVTVGYMYMLKLNHLVDDKMHARSTGSYSLVTQQPLGGKAQFGGQRFGEMEVWALEAYGAAYTLQEMLTVKSDDVNGRTKMYKNIVDGNHQMEPGMPESFNVLLKEIRSLGINIELEDE'\n",
    "corrected = uniprot_seq[151:224] + uniprot_seq[342:443]\n",
    "print(corrected)\n",
    "print(len(corrected))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d518c3c8-263a-43c7-9a15-2c3e552d363c",
   "metadata": {},
   "source": [
    "### Case Study 4 - "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b01d1d03-925d-4b68-8a2b-e84706074241",
   "metadata": {},
   "outputs": [],
   "source": [
    "given_seq = data[data['cath_id'] == '4dgwC00']['sequences'].item()\n",
    "print(given_seq)\n",
    "print(len(given_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "663c5ddd-bb68-4296-9192-75527e98ad86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This accession is CASE 2: Rows that have cath_indices but NaN sequences\n",
    "\n",
    "# PDB: 4dgw\n",
    "# cath_id: 4dgwC00\n",
    "# iloc: 4668\n",
    "# cath_indices: [(101, 253)]\t\t\n",
    "# sequences (given): NaN\n",
    "\n",
    "# It is missing many residues, and thus has a gap\n",
    "# [GAP DETECTED for cath_id 4dgwC00] Missing residues in range 101 - 253: [137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 153, 226]\n",
    "\n",
    "uniprot_seq = 'MNLLETRRSLLEEMEIIENAIAERIQRNPELYYHYIQESSKVFPDTKLPRSSLIAENKIYKFKKVKRKRKQIILQQHEINIFLRDYQEKQQTFNKINRPEETQEDDKDLPNFERKLQQLEKELKNEDENFELDINSKKDKYALFSSSSDPSRRTNILSDRARDLDLNEIFTRDEQYGEYMELEQFHSLWLNVIKRGDCSLLQFLDILELFLDDEKYLLTPPMDRKNDRYMAFLLKLSKYVETFFFKSYALLDAAAVENLIKSDFEHSYCRGSLRSEAKGIYCPFCSRWFKTSSVFESHLVGKIHKKNESKRRNFVYSEYKLHRYLKYLNDEFSRTRSFVERKLAFTANERMAEMDILTQKYEAPAYDSTEKEGAEQVDGEQRDGQLQEEHLSGKSFDMPLGPDGLPMPYWLYKLHGLDREYRCEICSNKVYNGRRTFERHFNEERHIYHLRCLGIEPSSVFKGITKIKEAQELWKNMQGQSQLTSIAAVPPKPNPSQLKVPTELELEEEDEEGNVMSKKVYDELKKQGLV'\n",
    "\n",
    "actual = uniprot_seq[100:253]\n",
    "print(actual)\n",
    "print(len(actual))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61404bbe-c944-4bbf-9fef-1f3a6aae1a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    # Data accession\n",
    "    pdb_code = '2gnx'\n",
    "    # cath_indices = [(101, 253)]\n",
    "\n",
    "    # Web scraping fxn calls\n",
    "    uniprot_accession = get_uniprot_accession_from_pdb(pdb_code)\n",
    "    fasta_sequence = get_fasta_sequence_from_uniprot(uniprot_accession)\n",
    "    print(fasta_sequence)\n",
    "    \n",
    "    # Concatenate the indexed fasta sequences if len(cath_indices) > 1\n",
    "    indexed_sequence = ''.join([fasta_sequence[start-1:end] for start, end in cath_indices]) \n",
    "\n",
    "    print(f\"Updated FASTA sequence for PDB code {pdb_code} (UniProt accession {uniprot_accession}):\\n{indexed_sequence}\\n\")\n",
    "    \n",
    "except Exception as e:\n",
    "    print(f\"Failed to fetch sequence for PDB code {pdb_code}: {str(e)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed4a2550-fb10-4a46-8a50-3e1ca1e3dd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from Bio.PDB.Polypeptide import protein_letters_3to1\n",
    "\n",
    "def get_sequence_from_pdb(pdb_filename):\n",
    "    \"\"\"\n",
    "    The sequences come from the PDB files\n",
    "    \"\"\"\n",
    "    pdb_parser = Bio.PDB.PDBParser()\n",
    "    structure = pdb_parser.get_structure(pdb_filename, pdb_filename)\n",
    "    assert len(structure) == 1\n",
    "\n",
    "    seq = []\n",
    "\n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            for residue in chain:\n",
    "                if residue.get_id()[0] == \" \":  # This checks if it's a standard residue\n",
    "                    try:\n",
    "                        seq.append(protein_letters_3to1[residue.get_resname()])\n",
    "                    except KeyError:\n",
    "                        seq.append('U')\n",
    "                else:\n",
    "                    print('nonstandard', residue.get_id())\n",
    "\n",
    "    return ''.join(seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8267115-5481-45ab-ab23-450d491bbcfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cath_id = '4dgwC00'\n",
    "\n",
    "# Open the PDB file\n",
    "pdb_filename = f\"../data/pdb_share/{cath_id}\"\n",
    "res = get_sequence_from_pdb(pdb_filename)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95260f7b-b816-43e4-93ce-8ce41f8baaef",
   "metadata": {},
   "outputs": [],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d96e6ea3-4635-43b1-8dad-38f6f93d8db8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66f41570-a7e0-4676-b45d-b9ba78c55211",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caa16d15-be3d-417b-bfb9-a28f76841a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolate data entries with sequence gaps\n",
    "gapped_data = data[data['gap'] == 1]\n",
    "gapped_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8745cfe-63a1-4ea3-bcfd-a2000b577d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load complete sequences from UniProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fb1af02-8018-4532-ae85-d507f28d9914",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edit gap sequences in the main data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5ae50e6-4cd8-49d4-85d9-c9af4794ac80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Confirm that all sequence gaps have been removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e95e2dc9-1286-4d22-ad58-0268055d6f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add sequence motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02d02f2c-6d22-459c-9be1-9fa3e284a4ad",
   "metadata": {},
   "source": [
    "# Structure Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742b2283-e178-4dc6-8c55-24e223461bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain and Store PDB files for non-gap sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "840173a6-6028-4bce-a4be-4e096bf2201b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain PDB files for gap sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a3b2c5-9bab-4f51-b529-f21c319ec920",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remodel using SCWRL4 and Store"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4c30448-49fe-4979-8fa7-68d73c26cc53",
   "metadata": {},
   "source": [
    "# Graph Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1023c659-8ca1-42bc-816c-6b46dd1579c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create NetworkX graphical representations of stored PDB files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec4b8c55-3600-4443-9f7c-c7eedba7de72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split data into train/test splits based on homological similarity"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
