{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da5aabfc-e467-46d4-a09a-6e8c89f5d684",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import *\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", message=\"Used element '.' for Atom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfb14584-e84d-42f7-8c60-077c7e380b6e",
   "metadata": {},
   "source": [
    "# Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f151586-70e4-447a-901c-71ba75943cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>(2, 40)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3r0hG01</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>2.60</td>\n",
       "      <td>3r0h</td>\n",
       "      <td>TAEIKPNKKILIELKVEKKPMGVIVCGGKNNHVTTGCVITHVYPEG...</td>\n",
       "      <td>[(479, 579)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1aqcA00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>2.30</td>\n",
       "      <td>1aqc</td>\n",
       "      <td>EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...</td>\n",
       "      <td>[(324, 488)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6hjfA02</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>310</td>\n",
       "      <td>10</td>\n",
       "      <td>1.70</td>\n",
       "      <td>6hjf</td>\n",
       "      <td>VPSFLYQQDVVVVLPKPYGEVRVDIAFGGNFFAIVPAEQLGIDISV...</td>\n",
       "      <td>[(189, 367)]</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1o1zA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>190</td>\n",
       "      <td>1.60</td>\n",
       "      <td>1o1z</td>\n",
       "      <td>HHHHVIVLGHRGYSAKYLENTLEAFMKAIEAGANGVELDVRLSKDG...</td>\n",
       "      <td>[(-3, 222)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cath_id  class  architecture  topology  superfamily  \\\n",
       "0  2w3sB01      3            90      1170           50   \n",
       "1  3be3A00      2            30        30          320   \n",
       "2  3zq4C03      3            10        20          580   \n",
       "3  1peqA03      1            10      1650           20   \n",
       "4  1bdoA00      2            40        50          100   \n",
       "5  3r0hG01      2            30        42           10   \n",
       "6  1aqcA00      2            30        29           30   \n",
       "7  2gnxA01      1            10      3450           30   \n",
       "8  6hjfA02      3            10       310           10   \n",
       "9  1o1zA00      3            20        20          190   \n",
       "\n",
       "   resolution_in_angstroms pdb_id  \\\n",
       "0                     2.60   2w3s   \n",
       "1                     2.04   3be3   \n",
       "2                     3.00   3zq4   \n",
       "3                     2.80   1peq   \n",
       "4                     1.80   1bdo   \n",
       "5                     2.60   3r0h   \n",
       "6                     2.30   1aqc   \n",
       "7                     2.45   2gnx   \n",
       "8                     1.70   6hjf   \n",
       "9                     1.60   1o1z   \n",
       "\n",
       "                                           sequences  \\\n",
       "0  SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1  QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPST...   \n",
       "2  DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVIN...   \n",
       "3  DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4  EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "5  TAEIKPNKKILIELKVEKKPMGVIVCGGKNNHVTTGCVITHVYPEG...   \n",
       "6  EDLIDGIIFAANYLGSTQLLSDKTPSKNVRQAQEAVSRIKAQKLTE...   \n",
       "7                                                NaN   \n",
       "8  VPSFLYQQDVVVVLPKPYGEVRVDIAFGGNFFAIVPAEQLGIDISV...   \n",
       "9  HHHHVIVLGHRGYSAKYLENTLEAFMKAIEAGANGVELDVRLSKDG...   \n",
       "\n",
       "               cath_indices architecture_id  label  \n",
       "0                [(2, 124)]         (3, 90)      9  \n",
       "1                 [(6, 81)]         (2, 30)      2  \n",
       "2              [(449, 555)]         (3, 10)      5  \n",
       "3              [(294, 346)]         (1, 10)      0  \n",
       "4               [(77, 156)]         (2, 40)      3  \n",
       "5              [(479, 579)]         (2, 30)      2  \n",
       "6              [(324, 488)]         (2, 30)      2  \n",
       "7  [(101, 295), (431, 440)]         (1, 10)      0  \n",
       "8              [(189, 367)]         (3, 10)      5  \n",
       "9               [(-3, 222)]         (3, 20)      6  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('../data/cath_w_seqs_share.csv', index_col=0)\n",
    "data['cath_indices'] = data['cath_indices'].apply(safe_eval) # Convert `cath_indices` from str to list of tuples\n",
    "\n",
    "# Map architecture IDs to names and labels\n",
    "architecture_names = {\n",
    "    (1, 10): {\"name\": \"Mainly Alpha: Orthogonal Bundle\", \"label\": 0},\n",
    "    (1, 20): {\"name\": \"Mainly Alpha: Up-down Bundle\", \"label\": 1},\n",
    "    (2, 30): {\"name\": \"Mainly Beta: Roll\", \"label\": 2},\n",
    "    (2, 40): {\"name\": \"Mainly Beta: Beta Barrel\", \"label\": 3},\n",
    "    (2, 60): {\"name\": \"Mainly Beta: Sandwich\", \"label\": 4},\n",
    "    (3, 10): {\"name\": \"Alpha Beta: Roll\", \"label\": 5},\n",
    "    (3, 20): {\"name\": \"Alpha Beta: Alpha-Beta Barrel\", \"label\": 6},\n",
    "    (3, 30): {\"name\": \"Alpha Beta: 2-Layer Sandwich\", \"label\": 7},\n",
    "    (3, 40): {\"name\": \"Alpha Beta: 3-Layer(aba) Sandwich\", \"label\": 8},\n",
    "    (3, 90): {\"name\": \"Alpha Beta: Alpha-Beta Complex\", \"label\": 9}\n",
    "}\n",
    "\n",
    "# Create 'architecture_id' and 'label' columns directly from architecture_names\n",
    "data['architecture_id'] = data.apply(lambda row: (row['class'], row['architecture']), axis=1)\n",
    "data['label'] = data.apply(lambda row: architecture_names[(row['class'], row['architecture'])]['label'], axis=1)\n",
    "\n",
    "\n",
    "data.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4216addd-f01b-4236-925a-ca43aa4568a6",
   "metadata": {},
   "source": [
    "# File Hierarchy Construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df668978-5c31-4fb0-9dc5-a5a5f67ec9b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in data.iterrows():\n",
    "    cath_id = row['cath_id']\n",
    "    \n",
    "    # Create ../data/{cath_id}/pdb directory if it doesn't exist\n",
    "    pdb_dir = f\"../data/{cath_id}/pdb\"\n",
    "    os.makedirs(pdb_dir, exist_ok=True)\n",
    "    \n",
    "    # Create ../data/{cath_id}/networkx directory if it doesn't exist\n",
    "    networkx_dir = f\"../data/{cath_id}/networkx\"\n",
    "    os.makedirs(networkx_dir, exist_ok=True)\n",
    "\n",
    "    # Create ../data/{cath_id}/networkx directory if it doesn't exist\n",
    "    seq_dir = f\"../data/{cath_id}/seqs\"\n",
    "    os.makedirs(seq_dir, exist_ok=True)\n",
    "    \n",
    "    # Copy file from ../data/pdb_share/{cath_id} to ../data/{cath_id}/pdb\n",
    "    source_file = f\"../data/pdb_share/{cath_id}\"\n",
    "    dest_file = f\"../data/{cath_id}/pdb/{cath_id}\"\n",
    "    shutil.copyfile(source_file, dest_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "efa31d8e-78d5-418b-8a45-5202ff82a554",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renumbering Alt Locs detected in PDB File 1qouB00:\n",
      "['32C', '32B', '94A', '129A']\n"
     ]
    }
   ],
   "source": [
    "cath_id = '1qouB00'\n",
    "dest_file = f\"../data/{cath_id}/pdb/{cath_id}\"\n",
    "\n",
    "# Remove amino acid ALT LOCs, should they occur in any destination pdb files\n",
    "with open(dest_file, 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "# Extract all residue identifiers\n",
    "resids = []\n",
    "for line in lines:\n",
    "    if line.startswith(\"ATOM\"):\n",
    "        resid = line[22:30].strip()\n",
    "        resids.append(resid)\n",
    "\n",
    "# Determine unique values with any alt locs\n",
    "unique_resids = sorted(set(resids), key=lambda x: int(x[:-1]) if x[-1].isalpha() else int(x))\n",
    "\n",
    "# Create a list of elements that end with a letter\n",
    "alt_locs = [resid for resid in unique_resids if resid[-1].isalpha()]\n",
    "\n",
    "# Check if any element ends with a letter\n",
    "contains_alt_loc = len(alt_locs) > 0\n",
    "\n",
    "if contains_alt_loc:\n",
    "    print(f'Renumbering Alt Locs detected in PDB File {cath_id}:\\n{alt_locs}')\n",
    "    corrected_resids = correct_alt_locs(dest_file, unique_resids)\n",
    "    renumber_pdb_file(corrected_resids, dest_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12348b1a-2104-493d-b21b-304e44b28ba8",
   "metadata": {},
   "source": [
    "# Handling Missing Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "343068da-533c-48a8-a85d-4d773423d736",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>19hcA01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.80</td>\n",
       "      <td>19hc</td>\n",
       "      <td>CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>2fd4A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>1.80</td>\n",
       "      <td>2fd4</td>\n",
       "      <td>KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>2xskA00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2420</td>\n",
       "      <td>1.70</td>\n",
       "      <td>2xsk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(3, 97)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3138</th>\n",
       "      <td>5inwA02</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>10</td>\n",
       "      <td>2.70</td>\n",
       "      <td>5inw</td>\n",
       "      <td>WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3838</th>\n",
       "      <td>1dkiC01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>70</td>\n",
       "      <td>50</td>\n",
       "      <td>1.60</td>\n",
       "      <td>1dki</td>\n",
       "      <td>FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4330</th>\n",
       "      <td>1kskA01</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>290</td>\n",
       "      <td>10</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1ksk</td>\n",
       "      <td>GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4668</th>\n",
       "      <td>4dgwC00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2690</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4dgw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 253)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>1nthA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>460</td>\n",
       "      <td>1.55</td>\n",
       "      <td>1nth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>3pieA02</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>1370</td>\n",
       "      <td>250</td>\n",
       "      <td>2.90</td>\n",
       "      <td>3pie</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(379, 487)]</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "7     2gnxA01      1            10      3450           30   \n",
       "178   19hcA01      3            90        10           10   \n",
       "2208  2fd4A00      3            30        40          110   \n",
       "2827  2xskA00      2            60        40         2420   \n",
       "3138  5inwA02      2            30        39           10   \n",
       "3838  1dkiC01      3            90        70           50   \n",
       "4330  1kskA01      3            10       290           10   \n",
       "4668  4dgwC00      2            60        40         2690   \n",
       "5776  1nthA00      3            20        20          460   \n",
       "6010  3pieA02      3            30      1370          250   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "7                        2.45   2gnx   \n",
       "178                      1.80   19hc   \n",
       "2208                     1.80   2fd4   \n",
       "2827                     1.70   2xsk   \n",
       "3138                     2.70   5inw   \n",
       "3838                     1.60   1dki   \n",
       "4330                     2.00   1ksk   \n",
       "4668                     3.11   4dgw   \n",
       "5776                     1.55   1nth   \n",
       "6010                     2.90   3pie   \n",
       "\n",
       "                                              sequences  \\\n",
       "7                                                   NaN   \n",
       "178   CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...   \n",
       "2208  KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...   \n",
       "2827                                                NaN   \n",
       "3138  WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...   \n",
       "3838  FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...   \n",
       "4330  GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...   \n",
       "4668                                                NaN   \n",
       "5776                                                NaN   \n",
       "6010                                                NaN   \n",
       "\n",
       "                  cath_indices architecture_id  label  \n",
       "7     [(101, 295), (431, 440)]         (1, 10)      0  \n",
       "178                        NaN         (3, 90)      9  \n",
       "2208                       NaN         (3, 30)      7  \n",
       "2827                 [(3, 97)]         (2, 60)      4  \n",
       "3138                       NaN         (2, 30)      2  \n",
       "3838                       NaN         (3, 90)      9  \n",
       "4330                       NaN         (3, 10)      5  \n",
       "4668              [(101, 253)]         (2, 60)      4  \n",
       "5776                [(2, 458)]         (3, 20)      6  \n",
       "6010              [(379, 487)]         (3, 30)      7  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify rows with any NaN values\n",
    "rows_with_nan = data[data.isna().any(axis=1)]\n",
    "rows_with_nan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57a1986f-45b4-453e-bdad-412a9a474ece",
   "metadata": {},
   "source": [
    "### CASE 1: Rows that have `sequences` but NaN `cath_indices` can be automatically corrected with residue indices present in the provided pdb files using a call to helper function `extract_resid_ranges()` which considers contiguous ranges of residues when assigning index tuples to `cath_indices`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "443e839e-9a51-4648-858b-5a81def74ad9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>19hcA01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.8</td>\n",
       "      <td>19hc</td>\n",
       "      <td>CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>2fd4A00</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>40</td>\n",
       "      <td>110</td>\n",
       "      <td>1.8</td>\n",
       "      <td>2fd4</td>\n",
       "      <td>KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3138</th>\n",
       "      <td>5inwA02</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>10</td>\n",
       "      <td>2.7</td>\n",
       "      <td>5inw</td>\n",
       "      <td>WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3838</th>\n",
       "      <td>1dkiC01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>70</td>\n",
       "      <td>50</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1dki</td>\n",
       "      <td>FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4330</th>\n",
       "      <td>1kskA01</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>290</td>\n",
       "      <td>10</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1ksk</td>\n",
       "      <td>GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "178   19hcA01      3            90        10           10   \n",
       "2208  2fd4A00      3            30        40          110   \n",
       "3138  5inwA02      2            30        39           10   \n",
       "3838  1dkiC01      3            90        70           50   \n",
       "4330  1kskA01      3            10       290           10   \n",
       "\n",
       "      resolution_in_angstroms pdb_id  \\\n",
       "178                       1.8   19hc   \n",
       "2208                      1.8   2fd4   \n",
       "3138                      2.7   5inw   \n",
       "3838                      1.6   1dki   \n",
       "4330                      2.0   1ksk   \n",
       "\n",
       "                                              sequences cath_indices  \\\n",
       "178   CHQSETKERRECAGCHAITTPKDDEAWCATTVAPVSPMLAPYKVVI...          NaN   \n",
       "2208  KLAALDPIASQFSQLRTISKALGFKDAADDVTHCLFGGELSLSNPD...          NaN   \n",
       "3138  WGNKFEPDLTKNVRFWVNSSYSMMVPTMHQRAKLSYTQDRKLRSTV...          NaN   \n",
       "3838  FARNEKEAKDSAITFIQKIKLDKVNLGGELSGSNMYVYNISTGGFV...          NaN   \n",
       "4330  GSHRLDKFIAQQLGVSRAIAGREIRGNRVTVDGEIVRNAAFKLLPE...          NaN   \n",
       "\n",
       "     architecture_id  label  \n",
       "178          (3, 90)      9  \n",
       "2208         (3, 30)      7  \n",
       "3138         (2, 30)      2  \n",
       "3838         (3, 90)      9  \n",
       "4330         (3, 10)      5  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case1 = data[data['sequences'].notna() & data['cath_indices'].isna()]\n",
    "case1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e60ddd3b-8843-45ba-aa07-62b2e4a9eb3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated `cath_indices` for cath_id 19hcA01:\n",
      "[(100, 129), (168, 292)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 2fd4A00:\n",
      "[(435, 553)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 5inwA02:\n",
      "[(251, 348), (408, 416)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 1dkiC01:\n",
      "[(1, 112), (154, 253)]\n",
      "\n",
      "Updated `cath_indices` for cath_id 1kskA01:\n",
      "[(1, 63)]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Define the threshold value for grouping contiguous residue ranges.\n",
    "Residue ranges are considered contiguous if the difference between consecutive residue IDs\n",
    "does not exceed this threshold value.\n",
    "This threshold value directly influences the identification of gaps in the next section, so it is important \n",
    "to choose this number carefully.\n",
    "'''\n",
    "gap_threshold = 20\n",
    "\n",
    "# Iterate over case1, apply get_sequence_from_pdb, and update data\n",
    "for index, row in case1.iterrows():\n",
    "    # Data accession\n",
    "    cath_id = row['cath_id']\n",
    "    pdb_filename = f\"../data/{cath_id}/pdb/{cath_id}\"\n",
    "\n",
    "    # Call provided helper fxn\n",
    "    cath_indices = extract_resid_ranges(pdb_filename, threshold=gap_threshold)\n",
    "\n",
    "    # Update the data frame to remove case2 NaN values\n",
    "    data.at[index, 'cath_indices'] = cath_indices\n",
    "    print(f\"Updated `cath_indices` for cath_id {cath_id}:\\n{cath_indices}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e405b4c6-1830-45fe-86e8-02b1fdcf904b",
   "metadata": {},
   "source": [
    "### CASE 2: Rows that have `cath_indices` but NaN `sequences` can be automatically corrected with the sequence present in UniProtKB\n",
    "- Note: Helper function `get_sequence_from_pdb()` does not suffice here because of the prescence of `UNK` resnames in case2 pdb files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "922472af-3e33-4263-ad08-20843e316cb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>2xskA00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2420</td>\n",
       "      <td>1.70</td>\n",
       "      <td>2xsk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(3, 97)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4668</th>\n",
       "      <td>4dgwC00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>40</td>\n",
       "      <td>2690</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4dgw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(101, 253)]</td>\n",
       "      <td>(2, 60)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5776</th>\n",
       "      <td>1nthA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>460</td>\n",
       "      <td>1.55</td>\n",
       "      <td>1nth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(2, 458)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>3pieA02</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>1370</td>\n",
       "      <td>250</td>\n",
       "      <td>2.90</td>\n",
       "      <td>3pie</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(379, 487)]</td>\n",
       "      <td>(3, 30)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cath_id  class  architecture  topology  superfamily  \\\n",
       "7     2gnxA01      1            10      3450           30   \n",
       "2827  2xskA00      2            60        40         2420   \n",
       "4668  4dgwC00      2            60        40         2690   \n",
       "5776  1nthA00      3            20        20          460   \n",
       "6010  3pieA02      3            30      1370          250   \n",
       "\n",
       "      resolution_in_angstroms pdb_id sequences              cath_indices  \\\n",
       "7                        2.45   2gnx       NaN  [(101, 295), (431, 440)]   \n",
       "2827                     1.70   2xsk       NaN                 [(3, 97)]   \n",
       "4668                     3.11   4dgw       NaN              [(101, 253)]   \n",
       "5776                     1.55   1nth       NaN                [(2, 458)]   \n",
       "6010                     2.90   3pie       NaN              [(379, 487)]   \n",
       "\n",
       "     architecture_id  label  \n",
       "7            (1, 10)      0  \n",
       "2827         (2, 60)      4  \n",
       "4668         (2, 60)      4  \n",
       "5776         (3, 20)      6  \n",
       "6010         (3, 30)      7  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case2 = data[data['sequences'].isna() & data['cath_indices'].notna()]\n",
    "case2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c13f123-c264-4c0a-bbb3-3a5c6592530d",
   "metadata": {},
   "source": [
    "# Handling Sequence Gaps\n",
    "- Phase 1: Detect Gaps\n",
    "- Phase 2: Replace Gaps in `sequences` col w/ AAs known by UniProtKB\n",
    "- Phase 3: Protein domain remodeling w/ SCWRL4\n",
    "  ##### Note: Case 2 data entries can be treated at the same time as sequence gaps because the NaN sequences for Case 2 entries, when filled from pdb_share, may still contain sequence gaps. The process for resolving Case 2 sequence gaps is the same for the larger data set. We have killed 2 birds with one stone!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "752eea20-377b-41fa-baea-587fc07214a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_cases = ['1qouB00']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cee007c5-173e-4c6e-8392-bd308a540c34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: string index out of range for CATH ID: 4dnnA00\n",
      "Error: string index out of range for CATH ID: 3hw5B00\n",
      "Error: string index out of range for CATH ID: 3uepA00\n",
      "Error: string index out of range for CATH ID: 3lmlA01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2gzqA00\n",
      "Error: string index out of range for CATH ID: 4lgdB02\n",
      "Error: string index out of range for CATH ID: 3k6tB00\n",
      "Error: write() argument must be str, not None for CATH ID: 3ayhA00\n",
      "Error: string index out of range for CATH ID: 2vdfA00\n",
      "Error: string index out of range for CATH ID: 1ukfA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 3g5tA00\n",
      "Error: string index out of range for CATH ID: 1xmxA01\n",
      "Error: string index out of range for CATH ID: 4iluA02\n",
      "Error: string index out of range for CATH ID: 3efmA02\n",
      "Error: string index out of range for CATH ID: 3bptA00\n",
      "Error: string index out of range for CATH ID: 3gmiA01\n",
      "Error: string index out of range for CATH ID: 3tigA01\n",
      "Error: More than one record found in handle for CATH ID: 3iylU02\n",
      "Error: string index out of range for CATH ID: 2atcB02\n",
      "Error: string index out of range for CATH ID: 2es9A00\n",
      "Error: string index out of range for CATH ID: 2y8yA01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2g2xA00\n",
      "Error: string index out of range for CATH ID: 2pmrA00\n",
      "Error: string index out of range for CATH ID: 5hdwA00\n",
      "Error: string index out of range for CATH ID: 2azpA01\n",
      "Error: string index out of range for CATH ID: 4nohA01\n",
      "Error: string index out of range for CATH ID: 4jejA00\n",
      "Error: string index out of range for CATH ID: 3no8A00\n",
      "Error: string index out of range for CATH ID: 1ii2A03\n",
      "Error: string index out of range for CATH ID: 2xi5A00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1m1jE01\n",
      "Error: string index out of range for CATH ID: 3g0tA01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 3g2eB00\n",
      "Error: string index out of range for CATH ID: 3t0pA01\n",
      "Error: string index out of range for CATH ID: 3hrgA01\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2aegA01\n",
      "Error: string index out of range for CATH ID: 1bowA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 3fz2A00\n",
      "Error: string index out of range for CATH ID: 1vq8V00\n",
      "Error: string index out of range for CATH ID: 4aajA00\n",
      "Error: string index out of range for CATH ID: 2rc3C00\n",
      "Error: string index out of range for CATH ID: 3go5A01\n",
      "Error: string index out of range for CATH ID: 4hhxA00\n",
      "Error: string index out of range for CATH ID: 2xliA01\n",
      "Error: string index out of range for CATH ID: 4eyyQ02\n",
      "Error: string index out of range for CATH ID: 3f5rA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 3l7vA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3cf6E01\n",
      "Error: string index out of range for CATH ID: 4mncA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1ztdA00\n",
      "Error: string index out of range for CATH ID: 3a9lA00\n",
      "Error: string index out of range for CATH ID: 3teeA02\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1w94A00\n",
      "Error: string index out of range for CATH ID: 3ik4A02\n",
      "Error: string index out of range for CATH ID: 3ggeA00\n",
      "Error: string index out of range for CATH ID: 4k59A00\n",
      "Error: string index out of range for CATH ID: 4az3B00\n",
      "Error: string index out of range for CATH ID: 2ovsA00\n",
      "Error: string index out of range for CATH ID: 4oelB00\n",
      "Error: string index out of range for CATH ID: 3devA02\n",
      "Error: string index out of range for CATH ID: 3q34A00\n",
      "Error: string index out of range for CATH ID: 3hl6A02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2x8kA01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1tr9A00\n",
      "Error: list index out of range for CATH ID: 3h74A00\n",
      "Error: string index out of range for CATH ID: 2pwwA00\n",
      "Error: string index out of range for CATH ID: 1yrvA00\n",
      "Error: More than one record found in handle for CATH ID: 4rmoA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1c1kA02\n",
      "Error: string index out of range for CATH ID: 2qiwA01\n",
      "Error: string index out of range for CATH ID: 3qyqA00\n",
      "Error: string index out of range for CATH ID: 2r8wA00\n",
      "Error: string index out of range for CATH ID: 1vpkA01\n",
      "Error: string index out of range for CATH ID: 3cbnA00\n",
      "Error: string index out of range for CATH ID: 2auaA01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2x3gA00\n",
      "Error: string index out of range for CATH ID: 2gfqA02\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1q9uA00\n",
      "Error: string index out of range for CATH ID: 1zq7A01\n",
      "Error: string index out of range for CATH ID: 4hdrB02\n",
      "Error: string index out of range for CATH ID: 1s2xA00\n",
      "Error: string index out of range for CATH ID: 4ivvA00\n",
      "Error: string index out of range for CATH ID: 4ebgA00\n",
      "Error: string index out of range for CATH ID: 1wcwA01\n",
      "Error: string index out of range for CATH ID: 6hjeB01\n",
      "Error: string index out of range for CATH ID: 1ynpB01\n",
      "Error: string index out of range for CATH ID: 2ph7A01\n",
      "Error: string index out of range for CATH ID: 2oaaB01\n",
      "Error: More than one record found in handle for CATH ID: 4hstB02\n",
      "Error: More than one record found in handle for CATH ID: 1r8oB01\n",
      "Error: string index out of range for CATH ID: 3mtqA00\n",
      "Error: string index out of range for CATH ID: 3d5pA00\n",
      "Error: string index out of range for CATH ID: 3df7A01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 3hx8A00\n",
      "Error: string index out of range for CATH ID: 3w5mA06\n",
      "Error: string index out of range for CATH ID: 1ki1B02\n",
      "Error: string index out of range for CATH ID: 3c0bA02\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2hd3A00\n",
      "Error: string index out of range for CATH ID: 4bhrA00\n",
      "Error: string index out of range for CATH ID: 3d4rE01\n",
      "Error: string index out of range for CATH ID: 2o9xA01\n",
      "Error: string index out of range for CATH ID: 4qqgG00\n",
      "Error: string index out of range for CATH ID: 4w8bA00\n",
      "Error: string index out of range for CATH ID: 2vlbC00\n",
      "Error: string index out of range for CATH ID: 1vpyA00\n",
      "Error: string index out of range for CATH ID: 4dunA01\n",
      "Error: string index out of range for CATH ID: 3a98A01\n",
      "Error: string index out of range for CATH ID: 1mw5A01\n",
      "Error: list index out of range for CATH ID: 3ht1A00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 5cofA00\n",
      "Error: string index out of range for CATH ID: 4esjA01\n",
      "Error: string index out of range for CATH ID: 2i9xA00\n",
      "Error: write() argument must be str, not None for CATH ID: 1zkrB00\n",
      "Error: string index out of range for CATH ID: 2otnB01\n",
      "Error: string index out of range for CATH ID: 3ihrA02\n",
      "Error: string index out of range for CATH ID: 4chmB00\n",
      "Error: write() argument must be str, not None for CATH ID: 1yq3C02\n",
      "Error: string index out of range for CATH ID: 3sl7A00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1xkrA00\n",
      "Error: string index out of range for CATH ID: 3uyiA01\n",
      "Error: string index out of range for CATH ID: 1t07A00\n",
      "Error: string index out of range for CATH ID: 2zeqA00\n",
      "Error: write() argument must be str, not None for CATH ID: 4kdiD00\n",
      "Error: write() argument must be str, not None for CATH ID: 3fjuB00\n",
      "Error: string index out of range for CATH ID: 3d3rB00\n",
      "Error: string index out of range for CATH ID: 3g74A00\n",
      "Error: string index out of range for CATH ID: 3fbgA01\n",
      "Error: string index out of range for CATH ID: 1ur1A00\n",
      "Error: string index out of range for CATH ID: 2wnvC00\n",
      "Error: string index out of range for CATH ID: 2bwfA00\n",
      "Error: string index out of range for CATH ID: 4da2A01\n",
      "Error: string index out of range for CATH ID: 1h10A00\n",
      "Error: string index out of range for CATH ID: 2f1vA00\n",
      "Error: string index out of range for CATH ID: 3b5mA01\n",
      "Error: string index out of range for CATH ID: 2dchX01\n",
      "Error: string index out of range for CATH ID: 1wz7A00\n",
      "Error: string index out of range for CATH ID: 3o6uA00\n",
      "Error: string index out of range for CATH ID: 2czrA01\n",
      "Error: string index out of range for CATH ID: 1vioA01\n",
      "Error: string index out of range for CATH ID: 2uv4A00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3cf6E03\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2a90A01\n",
      "Error: string index out of range for CATH ID: 1vloA04\n",
      "Error: string index out of range for CATH ID: 4f8lA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2o0pA00\n",
      "Error: string index out of range for CATH ID: 3gp6A00\n",
      "Error: string index out of range for CATH ID: 3jsbA01\n",
      "Error: string index out of range for CATH ID: 4ch7A02\n",
      "Error: string index out of range for CATH ID: 1rrmA02\n",
      "Error: string index out of range for CATH ID: 3rptA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 4ebrA00\n",
      "Error: More than one record found in handle for CATH ID: 4lflB00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 4c3iG02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2oqhA02\n",
      "Error: string index out of range for CATH ID: 2c4iA01\n",
      "Error: string index out of range for CATH ID: 1p4cA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2imrA02\n",
      "Error: string index out of range for CATH ID: 3beuA01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2o34A00\n",
      "Error: list index out of range for CATH ID: 1qw2A00\n",
      "Error: string index out of range for CATH ID: 1d5yB03\n",
      "Error: list index out of range for CATH ID: 1nc7A00\n",
      "Error: string index out of range for CATH ID: 4jz5A00\n",
      "Error: string index out of range for CATH ID: 3l9tA01\n",
      "Error: string index out of range for CATH ID: 3iuzA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1d0cA02\n",
      "Error: string index out of range for CATH ID: 2it9A00\n",
      "Error: string index out of range for CATH ID: 3mejA01\n",
      "Error: string index out of range for CATH ID: 2xvsA00\n",
      "Error: string index out of range for CATH ID: 2fl4A01\n",
      "Error: string index out of range for CATH ID: 3ooqA01\n",
      "Error: write() argument must be str, not None for CATH ID: 4i2yA01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/garmstrong/miniconda3/envs/medbio-challenge/lib/python3.11/site-packages/Bio/PDB/Atom.py:241: PDBConstructionWarning: Used element 'SE' for Atom (name=SE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: string index out of range for CATH ID: 2oogD00\n",
      "Error: string index out of range for CATH ID: 3l7iA01\n",
      "Error: string index out of range for CATH ID: 3bryA00\n",
      "Error: string index out of range for CATH ID: 3t4rA00\n",
      "Error: string index out of range for CATH ID: 2o6kA00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastp] Query is Empty!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: no element found: line 1, column 0 for CATH ID: 2ox5B00\n",
      "Error: write() argument must be str, not None for CATH ID: 4tpsD00\n",
      "Error: string index out of range for CATH ID: 2gdqA02\n",
      "Error: string index out of range for CATH ID: 5jheA01\n",
      "Error: string index out of range for CATH ID: 2hjmA01\n",
      "Error: string index out of range for CATH ID: 4fm3A00\n",
      "Error: write() argument must be str, not None for CATH ID: 4iqjM02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3l81A02\n",
      "Error: string index out of range for CATH ID: 3cygA02\n",
      "Error: string index out of range for CATH ID: 6fnuA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3lzdA01\n",
      "Error: string index out of range for CATH ID: 4fprB00\n",
      "Error: write() argument must be str, not None for CATH ID: 1h6wA03\n",
      "Error: string index out of range for CATH ID: 4b9gA00\n",
      "Error: string index out of range for CATH ID: 4p79A00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1y4cA03\n",
      "Error: string index out of range for CATH ID: 3s93A00\n",
      "Error: string index out of range for CATH ID: 3qyfA01\n",
      "Error: string index out of range for CATH ID: 3hbzA01\n",
      "Error: string index out of range for CATH ID: 5gkxA00\n",
      "Error: string index out of range for CATH ID: 1ig3A01\n",
      "Error: string index out of range for CATH ID: 3phuA01\n",
      "Error: string index out of range for CATH ID: 1vpqA00\n",
      "Error: write() argument must be str, not None for CATH ID: 1tfkB00\n",
      "Error: write() argument must be str, not None for CATH ID: 2b39A03\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 5kolD00\n",
      "Error: string index out of range for CATH ID: 3gvpA01\n",
      "Error: string index out of range for CATH ID: 3k2oA01\n",
      "Error: string index out of range for CATH ID: 1t7lB01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastp] Query is Empty!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: no element found: line 1, column 0 for CATH ID: 4bpxD00\n",
      "Error: string index out of range for CATH ID: 2cayB00\n",
      "Error: write() argument must be str, not None for CATH ID: 2b39A10\n",
      "Error: string index out of range for CATH ID: 1yleA02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3vz9B00\n",
      "Error: string index out of range for CATH ID: 3syyA00\n",
      "Error: string index out of range for CATH ID: 2cxhA01\n",
      "Error: string index out of range for CATH ID: 4gc3A00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3isrA01\n",
      "Error: string index out of range for CATH ID: 3lfjB00\n",
      "Error: More than one record found in handle for CATH ID: 3uwsA01\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2qejD01\n",
      "Error: string index out of range for CATH ID: 3c19A01\n",
      "Error: string index out of range for CATH ID: 1dtoA01\n",
      "Error: string index out of range for CATH ID: 4cbvA02\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1zn6A00\n",
      "Error: string index out of range for CATH ID: 3k59A01\n",
      "Error: string index out of range for CATH ID: 3dwoX00\n",
      "Error: string index out of range for CATH ID: 1qr0A01\n",
      "Error: write() argument must be str, not None for CATH ID: 2b39A13\n",
      "Error: string index out of range for CATH ID: 1o54A01\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 4ncbA01\n",
      "Error: string index out of range for CATH ID: 3a35A01\n",
      "Error: string index out of range for CATH ID: 1mjtB01\n",
      "Error: string index out of range for CATH ID: 3w1eA03\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3s6pD01\n",
      "Error: string index out of range for CATH ID: 5yyfC00\n",
      "Error: string index out of range for CATH ID: 3zmrB02\n",
      "Error: string index out of range for CATH ID: 3gkeA02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1tolA01\n",
      "Error: write() argument must be str, not None for CATH ID: 3f83A03\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2amxB00\n",
      "Error: string index out of range for CATH ID: 2wshA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1bh3A00\n",
      "Error: string index out of range for CATH ID: 1vmiA01\n",
      "Error: string index out of range for CATH ID: 1ohfC01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1n0eA00\n",
      "Error: write() argument must be str, not None for CATH ID: 1uuzB00\n",
      "Error: string index out of range for CATH ID: 1vhnA02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1b9mB02\n",
      "Error: string index out of range for CATH ID: 3oyyA01\n",
      "Error: string index out of range for CATH ID: 1u14A00\n",
      "Error: string index out of range for CATH ID: 3h0dA02\n",
      "Error: string index out of range for CATH ID: 1kskA01\n",
      "Error: string index out of range for CATH ID: 1okgA03\n",
      "Error: string index out of range for CATH ID: 3ai4A00\n",
      "Error: string index out of range for CATH ID: 3rbyA01\n",
      "Error: string index out of range for CATH ID: 4g09A03\n",
      "Error: string index out of range for CATH ID: 4bgpA01\n",
      "Error: string index out of range for CATH ID: 1rocA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 4ntqB00\n",
      "Error: string index out of range for CATH ID: 4njcA00\n",
      "Error: string index out of range for CATH ID: 3k6qA01\n",
      "Error: More than one record found in handle for CATH ID: 3akoC00\n",
      "Error: string index out of range for CATH ID: 4kruA00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/garmstrong/miniconda3/envs/medbio-challenge/lib/python3.11/site-packages/Bio/PDB/StructureBuilder.py:282: PDBConstructionWarning: WARNING: disordered atom found with blank altloc before line 817.\n",
      "\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: More than one record found in handle for CATH ID: 6ijfC01\n",
      "Error: string index out of range for CATH ID: 2gwnA01\n",
      "Error: string index out of range for CATH ID: 3nkzA00\n",
      "Error: More than one record found in handle for CATH ID: 3iylB03\n",
      "Error: string index out of range for CATH ID: 1acxA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 4heaA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1s4cC00\n",
      "Error: string index out of range for CATH ID: 4beuA01\n",
      "Error: string index out of range for CATH ID: 2xxpA02\n",
      "Error: string index out of range for CATH ID: 2d00A01\n",
      "Error: string index out of range for CATH ID: 4ub9A01\n",
      "Error: string index out of range for CATH ID: 3ipjA01\n",
      "Error: write() argument must be str, not None for CATH ID: 3aqbA00\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 1ameA00\n",
      "Error: string index out of range for CATH ID: 3am2A02\n",
      "Error: More than one record found in handle for CATH ID: 4hstA02\n",
      "Error: string index out of range for CATH ID: 3wscA00\n",
      "Error: string index out of range for CATH ID: 1vi7A01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 3d8lA00\n",
      "Error: string index out of range for CATH ID: 4kh9A01\n",
      "Error: string index out of range for CATH ID: 3gg7A00\n",
      "Error: string index out of range for CATH ID: 2yb1A01\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 1mjtA02\n",
      "Error: string index out of range for CATH ID: 1xrtB01\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2nyhA00\n",
      "Error: string index out of range for CATH ID: 5hp5A01\n",
      "Error: string index out of range for CATH ID: 4o1eB00\n",
      "Error: string index out of range for CATH ID: 1zs7A01\n",
      "Error: string index out of range for CATH ID: 2gfqA01\n",
      "Error: string index out of range for CATH ID: 2hiyA01\n",
      "Error: string index out of range for CATH ID: 3kkgA00\n",
      "Error: string index out of range for CATH ID: 2vt8A00\n",
      "Error: string index out of range for CATH ID: 3mswA00\n",
      "Error: string index out of range for CATH ID: 3hhsB03\n",
      "Error: string index out of range for CATH ID: 4ak2A02\n",
      "Error: string index out of range for CATH ID: 4cgsA00\n",
      "Error: string index out of range for CATH ID: 1vlpA00\n",
      "Error: string index out of range for CATH ID: 3gbyA00\n",
      "Error: string index out of range for CATH ID: 2f1cX00\n",
      "Error: string index out of range for CATH ID: 5trdA02\n",
      "Error: string index out of range for CATH ID: 1vk6A01\n",
      "Error: string index out of range for CATH ID: 5w17A01\n",
      "Error: string index out of range for CATH ID: 3payA01\n",
      "Error: string index out of range for CATH ID: 2c71A00\n",
      "Error: string index out of range for CATH ID: 2ynkA00\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2j7qA00\n",
      "Error: string index out of range for CATH ID: 2etdA00\n",
      "Error: string index out of range for CATH ID: 3hp7A01\n",
      "Error: string index out of range for CATH ID: 2hb0B02\n",
      "Error: string index out of range for CATH ID: 2wyoC01\n",
      "Error: string index out of range for CATH ID: 2a2jA00\n",
      "Error: More than one record found in handle for CATH ID: 3iylW03\n",
      "Error: cannot access local variable 'modified' where it is not associated with a value for CATH ID: 2gk4A00\n",
      "Error: string index out of range for CATH ID: 2r6vA01\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3dupA02\n",
      "Error: string index out of range for CATH ID: 4fqeA00\n",
      "Error: string index out of range for CATH ID: 3ks5A00\n",
      "Error: string index out of range for CATH ID: 4f78A01\n",
      "Error: string index out of range for CATH ID: 3hxlA05\n",
      "Error: string index out of range for CATH ID: 1jcjA00\n",
      "Error: string index out of range for CATH ID: 1j6rA00\n",
      "Error: string index out of range for CATH ID: 3dtdD00\n",
      "Error: string index out of range for CATH ID: 3zbdA00\n",
      "Error: string index out of range for CATH ID: 3u9rB01\n",
      "Error: string index out of range for CATH ID: 4aweA00\n",
      "Error: string index out of range for CATH ID: 2vbuA01\n",
      "Error: string index out of range for CATH ID: 1ym5A01\n",
      "Error: More than one record found in handle for CATH ID: 3iylW01\n",
      "Error: string index out of range for CATH ID: 4rdbA01\n",
      "Error: string index out of range for CATH ID: 5d3xB00\n",
      "Error: string index out of range for CATH ID: 2x8fA02\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 3er9B03\n",
      "Error: string index out of range for CATH ID: 3zzoA00\n",
      "Error: string index out of range for CATH ID: 2npnA01\n",
      "Error: string index out of range for CATH ID: 3ru6B00\n",
      "Error: string index out of range for CATH ID: 2gjwC01\n",
      "Error: string index out of range for CATH ID: 4m70I00\n",
      "Error: write() argument must be str, not None for CATH ID: 2fhzA00\n",
      "Error: string index out of range for CATH ID: 3b79A00\n",
      "Error: string index out of range for CATH ID: 2igsA00\n",
      "Error: string index out of range for CATH ID: 3qdfA01\n",
      "Error: string index out of range for CATH ID: 3fw3B00\n",
      "Error: string index out of range for CATH ID: 2in5A00\n",
      "Error: string index out of range for CATH ID: 4mwaA00\n",
      "Error: list index out of range for CATH ID: 1bqcA00\n",
      "Error: string index out of range for CATH ID: 2xciC01\n",
      "Error: list index out of range for CATH ID: 2xq2A01\n",
      "Error: string index out of range for CATH ID: 3qfwB02\n",
      "Error: string index out of range for CATH ID: 1xkgA00\n",
      "Error: string index out of range for CATH ID: 2pn5A10\n",
      "Error: object of type 'NoneType' has no len() for CATH ID: 2b4vA02\n",
      "Error: string index out of range for CATH ID: 4b9iA00\n",
      "Error: string index out of range for CATH ID: 1x8mA01\n",
      "Error: string index out of range for CATH ID: 2i71A01\n",
      "Error: string index out of range for CATH ID: 1tqgA00\n",
      "Error: string index out of range for CATH ID: 1xl4A02\n",
      "Error: string index out of range for CATH ID: 3wfdB00\n",
      "Error: string index out of range for CATH ID: 2qf4A01\n"
     ]
    }
   ],
   "source": [
    "duplicate_pdb_ids = set(data.groupby('pdb_id').filter(lambda x: len(x) > 1)['pdb_id']) # There are 720 PDB IDs with duplicate entries\n",
    "\n",
    "with open('data_processing_main_loop_output.txt', 'w') as output_file: # Open a file for all output to reduce I/O burden\n",
    "    for index, row in data.iterrows():\n",
    "        try:\n",
    "            result = []\n",
    "            \n",
    "            # Data accession\n",
    "            pdb_code = row['pdb_id']\n",
    "            cath_id = row['cath_id']\n",
    "            cath_indices = row['cath_indices']\n",
    "            given_seq = row['sequences']\n",
    "            print(f'[CATH_ID] {cath_id} ({index+1} / {len(data)})\\n', file=output_file)\n",
    "            \n",
    "            # Web scraping fxn calls (UniProtKB and RCSB PDB)\n",
    "            try:\n",
    "                uniprot_accession = get_uniprot_accession_from_pdb(pdb_code, duplicate_pdb_ids, output_file)\n",
    "                if uniprot_accession:\n",
    "                    # `uniprot_accession` is returned as a list from `get_uniprot_accession_from_pdb` if \n",
    "                    # 1) Your `pdb_code` reoccurs throughout the data set (duplicates), and multiple UniProt IDs are associated with it. \n",
    "                    #    This is logical for well studied proteins.\n",
    "                    # 2) [EDGE CASE] Your `pdb_code` is unique in the data set, but multiple UniProt IDs are associated with it.\n",
    "                    if isinstance(uniprot_accession, list): \n",
    "                        print(f'Identifying the correct subject sequence for [CATH ID] {cath_id}.\\n', file=output_file)\n",
    "                        if len(uniprot_accession) == 1: # 1 Uniprot ID represents mutltiple data entries\n",
    "                            uniprot_accession = uniprot_accession[0] # We pass a str to the function\n",
    "                            subject_sequence, _ = get_fasta_sequence_from_uniprot(uniprot_accession, output_file) \n",
    "                        else: \n",
    "                            # PDB ID is member of duplicate_pdb_ids with more than 1 possible Uniprot ID. \n",
    "                            # Filter through sequences for all Uniprot IDs to select the best match.\n",
    "                            # `uniprot_accession` is a list passed to the function. We return the seq and ID associated with the filtered match\n",
    "                            subject_sequence, uniprot_accession = get_fasta_sequence_from_uniprot(uniprot_accession, output_file, cath_id=cath_id, duplicates=True)\n",
    "                    else:\n",
    "                        # Here, `uniprot_accession` is of type str because only one UniprotID is associated with the `pdb_code`\n",
    "                        subject_sequence, _ = get_fasta_sequence_from_uniprot(uniprot_accession, output_file)\n",
    "                else:\n",
    "                    raise ValueError(f\"[EDGE CASE] UniProtKB does not contain data for PDB code {pdb_code}\\n\")\n",
    "    \n",
    "                print(f'Subject:  (UniProt accession {uniprot_accession})\\n{subject_sequence}\\n', file=output_file)\n",
    "                data.at[index, 'uniprot_id'] = uniprot_accession\n",
    "    \n",
    "                    \n",
    "            # Plan B: RCSB Protein Data Bank\n",
    "            except (requests.exceptions.HTTPError, ValueError):\n",
    "                print(f\"Trying to get FASTA sequence for PDB code {pdb_code} from RCSB PDB...\", file=output_file)\n",
    "                subject_sequence = get_fasta_sequence_from_rcsb(pdb_code, output_file)\n",
    "                if subject_sequence:\n",
    "                    print(f'Subject: (RCSB PDB)\\n{subject_sequence}\\n', file=output_file)\n",
    "                    data.at[index, 'uniprot_id'] = 'RCSB Lookup'\n",
    "                else:\n",
    "                    print(f\"Failed to retrieve both UniProt accession and FASTA sequence for PDB code {pdb_code}.\", file=output_file)\n",
    "    \n",
    "            # Write the `subject_sequence` to a file for use by NCBI BLAST+\n",
    "            subject_file = f'../data/{cath_id}/seqs/subject.fasta'\n",
    "            with open(subject_file, 'w') as t:\n",
    "                t.write(subject_sequence)\n",
    "            \n",
    "            if pd.isna(given_seq):    \n",
    "                print(f'Given Seq:\\nCase 2 - {given_seq}\\nBuilding query sequence(s) from pdb_share...\\n', file=output_file)\n",
    "            else:\n",
    "                print(f'Given Seq:\\n{given_seq}\\n', file=output_file)\n",
    "                # The `given_seq`, when present in the data frame, may consist of multiple protein domains \n",
    "                # which are treated separately in the folllowing for loop by enumerating the `cath_indices`\n",
    "            \n",
    "            pdb_filename = f\"../data/{cath_id}/pdb/{cath_id}\"\n",
    "            for i, idx_range in enumerate(cath_indices):\n",
    "                # The query sequence corresponds to the protein domain specific to the idx_range tuple\n",
    "                query_sequence = get_domain_sequence_from_pdb(pdb_filename, idx_range, cath_id, output_file) \n",
    "                print(f'\\nQuery Sequence {i+1}:  {idx_range}\\n{query_sequence}\\n', file=output_file)\n",
    "                query_file = f'../data/{cath_id}/seqs/query_{i+1}.fasta'\n",
    "                with open(query_file, 'w') as q:\n",
    "                    q.write(query_sequence)\n",
    "            \n",
    "                # Perform sequence alignment for this domain segment using NCBI Blast+\n",
    "                response_text = f'../data/{cath_id}/seqs/output_{i+1}.txt'\n",
    "                aligned_segment = extract_aligned_segment(query_sequence, subject_sequence, query_file, subject_file, response_text)\n",
    "                print(f'Alignment Segment (from subject):\\n{aligned_segment}\\n', file=output_file)\n",
    "    \n",
    "                if cath_id not in edge_cases:\n",
    "                    # Error Handling - Sequence alignment w/ BLAST+ sometimes leaves out amino acids. If not handled, this will corrupt remodeling.\n",
    "                    # However, certain edge_cases do not need this step because the aligned segment is already the correct result\n",
    "                    expected_num_AAs = idx_range[1] - idx_range[0] + 1\n",
    "                    if len(aligned_segment) != expected_num_AAs:\n",
    "                        print(f\"WARNING: Length of aligned segment [{len(aligned_segment)}] does not match the expected number of amino acids [{expected_num_AAs}]\\n\", file=output_file)\n",
    "                        # Manual sequence alignment\n",
    "                        aligned_segment = manual_alignment(query_sequence, aligned_segment, subject_sequence, output_file)\n",
    "            \n",
    "                result.append(aligned_segment)\n",
    "            \n",
    "            \n",
    "            # Concatenate the indexed fasta sequences if len(cath_indices) > 1\n",
    "            updated_sequence = ''.join(result)\n",
    "            \n",
    "            # Update the data frame to 1) Remove Case2 NaN values and 2) Handle Sequence Gaps\n",
    "            data.at[index, 'sequences'] = updated_sequence\n",
    "            print('[REPAIRED FULL SEQUENCE]', file=output_file)\n",
    "            print(f\"Updated FASTA sequence for CATH ID {cath_id} :\\n{updated_sequence}\\n\", file=output_file)\n",
    "        \n",
    "        except Exception as e:\n",
    "            with open('error_log.txt', 'a') as log_file:\n",
    "                log_file.write(f\"Error: {str(e)} for CATH ID: {cath_id}\\n\")\n",
    "            \n",
    "            # Optionally print the error to the console as well\n",
    "            print(f\"Error: {str(e)} for CATH ID: {cath_id}\")\n",
    "            # raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cb582a65-b383-4421-b1f1-1e8dabc77e43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cath_id</th>\n",
       "      <th>class</th>\n",
       "      <th>architecture</th>\n",
       "      <th>topology</th>\n",
       "      <th>superfamily</th>\n",
       "      <th>resolution_in_angstroms</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>sequences</th>\n",
       "      <th>cath_indices</th>\n",
       "      <th>architecture_id</th>\n",
       "      <th>label</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2w3sB01</td>\n",
       "      <td>3</td>\n",
       "      <td>90</td>\n",
       "      <td>1170</td>\n",
       "      <td>50</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2w3s</td>\n",
       "      <td>SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...</td>\n",
       "      <td>[(2, 124)]</td>\n",
       "      <td>(3, 90)</td>\n",
       "      <td>9</td>\n",
       "      <td>O54051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3be3A00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>320</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3be3</td>\n",
       "      <td>QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPMS...</td>\n",
       "      <td>[(6, 81)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "      <td>A0A0H3LS10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3zq4C03</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>580</td>\n",
       "      <td>3.00</td>\n",
       "      <td>3zq4</td>\n",
       "      <td>DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVYM...</td>\n",
       "      <td>[(449, 555)]</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "      <td>Q45493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1peqA03</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1650</td>\n",
       "      <td>20</td>\n",
       "      <td>2.80</td>\n",
       "      <td>1peq</td>\n",
       "      <td>DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...</td>\n",
       "      <td>[(294, 346)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "      <td>Q08698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1bdoA00</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1bdo</td>\n",
       "      <td>EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...</td>\n",
       "      <td>[(77, 156)]</td>\n",
       "      <td>(2, 40)</td>\n",
       "      <td>3</td>\n",
       "      <td>P0ABD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3r0hG01</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>2.60</td>\n",
       "      <td>3r0h</td>\n",
       "      <td>TAEIKPNKKILIELKVEKKPMGVIVCGGKNNHVTTGCVITHVYPEG...</td>\n",
       "      <td>[(479, 579)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "      <td>Q24008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1aqcA00</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>2.30</td>\n",
       "      <td>1aqc</td>\n",
       "      <td>EDLIDGIIFAANYLGSTQLLSDKTPSKNVRMMQAQEAVSRIKMAQK...</td>\n",
       "      <td>[(324, 488)]</td>\n",
       "      <td>(2, 30)</td>\n",
       "      <td>2</td>\n",
       "      <td>Q02410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2gnxA01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>3450</td>\n",
       "      <td>30</td>\n",
       "      <td>2.45</td>\n",
       "      <td>2gnx</td>\n",
       "      <td>LKKVVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKM...</td>\n",
       "      <td>[(101, 295), (431, 440)]</td>\n",
       "      <td>(1, 10)</td>\n",
       "      <td>0</td>\n",
       "      <td>Q6P1I3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6hjfA02</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>310</td>\n",
       "      <td>10</td>\n",
       "      <td>1.70</td>\n",
       "      <td>6hjf</td>\n",
       "      <td>VPSFLYQQDVVVVLPKPYGEVRVDIAFGGNFFAIVPAEQLGIDISV...</td>\n",
       "      <td>[(189, 367)]</td>\n",
       "      <td>(3, 10)</td>\n",
       "      <td>5</td>\n",
       "      <td>Q4DA80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1o1zA00</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>190</td>\n",
       "      <td>1.60</td>\n",
       "      <td>1o1z</td>\n",
       "      <td>HHHHMIVLGHRGYSAKYLENTLEAFMKAIEAGANGVELDVRLSKDG...</td>\n",
       "      <td>[(-3, 222)]</td>\n",
       "      <td>(3, 20)</td>\n",
       "      <td>6</td>\n",
       "      <td>Q9X1V6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cath_id  class  architecture  topology  superfamily  \\\n",
       "0  2w3sB01      3            90      1170           50   \n",
       "1  3be3A00      2            30        30          320   \n",
       "2  3zq4C03      3            10        20          580   \n",
       "3  1peqA03      1            10      1650           20   \n",
       "4  1bdoA00      2            40        50          100   \n",
       "5  3r0hG01      2            30        42           10   \n",
       "6  1aqcA00      2            30        29           30   \n",
       "7  2gnxA01      1            10      3450           30   \n",
       "8  6hjfA02      3            10       310           10   \n",
       "9  1o1zA00      3            20        20          190   \n",
       "\n",
       "   resolution_in_angstroms pdb_id  \\\n",
       "0                     2.60   2w3s   \n",
       "1                     2.04   3be3   \n",
       "2                     3.00   3zq4   \n",
       "3                     2.80   1peq   \n",
       "4                     1.80   1bdo   \n",
       "5                     2.60   3r0h   \n",
       "6                     2.30   1aqc   \n",
       "7                     2.45   2gnx   \n",
       "8                     1.70   6hjf   \n",
       "9                     1.60   1o1z   \n",
       "\n",
       "                                           sequences  \\\n",
       "0  SVGKPLPHDSARAHVTGQARYLDDLPCPANTLHLAFGLSTEASAAI...   \n",
       "1  QDFRPGVYRHYKGDHYLALGLARADETDEVVVVYTRLYARAGLPMS...   \n",
       "2  DIGNIVLRDRRILSEEGLVIVVVSIDMDDFKISAGPDLISRGFVYM...   \n",
       "3  DITFRLAKENAQMALFSPYDIQRRYGKPFGDIAISERYDELIADPH...   \n",
       "4  EISGHIVRSPMVGTFYRTPSPDAKAFIEVGQKVNVGDTLCIVEAMK...   \n",
       "5  TAEIKPNKKILIELKVEKKPMGVIVCGGKNNHVTTGCVITHVYPEG...   \n",
       "6  EDLIDGIIFAANYLGSTQLLSDKTPSKNVRMMQAQEAVSRIKMAQK...   \n",
       "7  LKKVVAGRGAPGGTAPHVEELLPHLSEQLCFFVQARMEIADFYEKM...   \n",
       "8  VPSFLYQQDVVVVLPKPYGEVRVDIAFGGNFFAIVPAEQLGIDISV...   \n",
       "9  HHHHMIVLGHRGYSAKYLENTLEAFMKAIEAGANGVELDVRLSKDG...   \n",
       "\n",
       "               cath_indices architecture_id  label  uniprot_id  \n",
       "0                [(2, 124)]         (3, 90)      9      O54051  \n",
       "1                 [(6, 81)]         (2, 30)      2  A0A0H3LS10  \n",
       "2              [(449, 555)]         (3, 10)      5      Q45493  \n",
       "3              [(294, 346)]         (1, 10)      0      Q08698  \n",
       "4               [(77, 156)]         (2, 40)      3      P0ABD8  \n",
       "5              [(479, 579)]         (2, 30)      2      Q24008  \n",
       "6              [(324, 488)]         (2, 30)      2      Q02410  \n",
       "7  [(101, 295), (431, 440)]         (1, 10)      0      Q6P1I3  \n",
       "8              [(189, 367)]         (3, 10)      5      Q4DA80  \n",
       "9               [(-3, 222)]         (3, 20)      6      Q9X1V6  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6fa7fa55-152e-4704-9e92-9796474f13cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered data has been saved to 'cleaned_data.csv'.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Open the error_log.txt file and read its content\n",
    "with open('error_log.txt', 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "# Extract the last 7 characters of each line\n",
    "cath_id_error_list = [line[-8:].strip() for line in lines]\n",
    "\n",
    "# Obtain the unique members from this list\n",
    "unique = set(cath_id_error_list)\n",
    "\n",
    "# Remove entries from the DataFrame if their 'cath_id' is in the set of unique members\n",
    "filtered_data = data[~data['cath_id'].isin(unique)]\n",
    "\n",
    "# Save the filtered DataFrame to a CSV file\n",
    "filtered_data.to_csv('cleaned_data.csv', index=False)\n",
    "\n",
    "# Checkpoint\n",
    "print(\"[CHECKPOINT] Filtered data has been saved to 'cleaned_data.csv'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a66c0981-faa5-44a7-9dc0-a9d2e9ac76d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# duplicate_pdb_ids = set(data.groupby('pdb_id').filter(lambda x: len(x) > 1)['pdb_id']) # There are 720 PDB IDs with duplicate entries\n",
    "\n",
    "# try:\n",
    "#     result = []\n",
    "    \n",
    "#     # Data accession\n",
    "#     pdb_code = '3uep'\n",
    "#     cath_id = '3uepA00'\n",
    "#     cath_indices = [(229, 313)]\t\n",
    "#     given_seq = 'PLTDLNQLPVQVSFEVGRQILDWHTLTSLEPGSLIDLTTPVDGEVRLLANGRLLGHGRLVEIQGRLGVRIERLTEVTISLEVLFQ'\n",
    "#     print(f'[CATH_ID] {cath_id}\\n')\n",
    "    \n",
    "#     # Web scraping fxn calls (UniProtKB and RCSB PDB)\n",
    "#     try:\n",
    "#         uniprot_accession = get_uniprot_accession_from_pdb(pdb_code, duplicate_pdb_ids)\n",
    "#         if uniprot_accession:\n",
    "#             # `uniprot_accession` is returned as a list from `get_uniprot_accession_from_pdb` if \n",
    "#             # 1) Your `pdb_code` reoccurs throughout the data set (duplicates), and multiple UniProt IDs are associated with it. \n",
    "#             #    This is logical for well studied proteins.\n",
    "#             # 2) [EDGE CASE] Your `pdb_code` is unique in the data set, but multiple UniProt IDs are associated with it.\n",
    "#             if isinstance(uniprot_accession, list): \n",
    "#                 print(f'Identifying the correct subject sequence for [CATH ID] {cath_id}.\\n')\n",
    "#                 if len(uniprot_accession) == 1: # 1 Uniprot ID represents mutltiple data entries\n",
    "#                     uniprot_accession = uniprot_accession[0] # We pass a str to the function\n",
    "#                     subject_sequence, _ = get_fasta_sequence_from_uniprot(uniprot_accession) \n",
    "#                 else: \n",
    "#                     # PDB ID is member of duplicate_pdb_ids with more than 1 possible Uniprot ID. \n",
    "#                     # Filter through sequences for all Uniprot IDs to select the best match.\n",
    "#                     # `uniprot_accession` is a list passed to the function. We return the seq and ID associated with the filtered match\n",
    "#                     subject_sequence, uniprot_accession = get_fasta_sequence_from_uniprot(uniprot_accession, cath_id=cath_id, duplicates=True)\n",
    "#             else:\n",
    "#                 # Here, `uniprot_accession` is of type str because only one UniprotID is associated with the `pdb_code`\n",
    "#                 subject_sequence, _ = get_fasta_sequence_from_uniprot(uniprot_accession)\n",
    "#         else:\n",
    "#             raise ValueError(f\"[EDGE CASE] UniProtKB does not contain data for PDB code {pdb_code}\\n\")\n",
    "\n",
    "#         print(f'Subject:  (UniProt accession {uniprot_accession})\\n{subject_sequence}\\n')\n",
    "\n",
    "            \n",
    "#     # Plan B: RCSB Protein Data Bank\n",
    "#     except (requests.exceptions.HTTPError, ValueError):\n",
    "#         print(f\"Trying to get FASTA sequence for PDB code {pdb_code} from RCSB PDB...\")\n",
    "#         subject_sequence = get_fasta_sequence_from_rcsb(pdb_code)\n",
    "#         if subject_sequence:\n",
    "#             print(f'Subject: (RCSB PDB)\\n{subject_sequence}\\n')\n",
    "#         else:\n",
    "#             print(f\"Failed to retrieve both UniProt accession and FASTA sequence for PDB code {pdb_code}.\")\n",
    "\n",
    "#     # Write the `subject_sequence` to a file for use by NCBI BLAST+\n",
    "#     subject_file = f'../data/{cath_id}/seqs/subject.fasta'\n",
    "#     with open(subject_file, 'w') as t:\n",
    "#         t.write(subject_sequence)\n",
    "    \n",
    "#     if pd.isna(given_seq):    \n",
    "#         print(f'Given Seq:\\nCase 2 - {given_seq}\\nBuilding query sequence(s) from pdb_share...\\n')\n",
    "#     else:\n",
    "#         print(f'Given Seq:\\n{given_seq}\\n')\n",
    "#         # The `given_seq`, when present in the data frame, may consist of multiple protein domains \n",
    "#         # which are treated separately in the folllowing for loop by enumerating the `cath_indices`\n",
    "    \n",
    "#     pdb_filename = f\"../data/{cath_id}/pdb/{cath_id}\"\n",
    "#     for i, idx_range in enumerate(cath_indices):\n",
    "#         # The query sequence corresponds to the protein domain specific to the idx_range tuple\n",
    "#         query_sequence = get_domain_sequence_from_pdb(pdb_filename, idx_range, cath_id) \n",
    "#         print(f'\\nQuery Sequence {i+1}:  {idx_range}\\n{query_sequence}\\n')\n",
    "#         query_file = f'../data/{cath_id}/seqs/query_{i+1}.fasta'\n",
    "#         with open(query_file, 'w') as q:\n",
    "#             q.write(query_sequence)\n",
    "    \n",
    "#         # Perform sequence alignment for this domain segment using NCBI Blast+\n",
    "#         output_file = f'../data/{cath_id}/seqs/output_{i+1}.txt'\n",
    "#         aligned_segment = extract_aligned_segment(query_sequence, subject_sequence, query_file, subject_file, output_file)\n",
    "#         print(f'Alignment Segment (from subject):\\n{aligned_segment}\\n')\n",
    "\n",
    "#         if cath_id not in edge_cases:\n",
    "#             # Error Handling - Sequence alignment w/ BLAST+ sometimes leaves out amino acids. If not handled, this will corrupt remodeling.\n",
    "#             # However, certain edge_cases do not need this step because the aligned segment is already the correct result\n",
    "#             expected_num_AAs = idx_range[1] - idx_range[0] + 1\n",
    "#             if len(aligned_segment) != expected_num_AAs:\n",
    "#                 print(f\"WARNING: Length of aligned segment [{len(aligned_segment)}] does not match the expected number of amino acids [{expected_num_AAs}]\\n\")\n",
    "#                 # Manual sequence alignment\n",
    "#                 aligned_segment = manual_alignment(query_sequence, aligned_segment, subject_sequence, output_file)\n",
    "    \n",
    "#         result.append(aligned_segment)\n",
    "    \n",
    "    \n",
    "#     # Concatenate the indexed fasta sequences if len(cath_indices) > 1\n",
    "#     updated_sequence = ''.join(result)\n",
    "    \n",
    "#     # Update the data frame to 1) Remove Case2 NaN values and 2) Handle Sequence Gaps\n",
    "#     data.at[index, 'sequences'] = updated_sequence\n",
    "#     print('[REPAIRED FULL SEQUENCE]')\n",
    "#     print(f\"Updated FASTA sequence for CATH ID {cath_id} :\\n{updated_sequence}\\n\")\n",
    "\n",
    "# except Exception as e:\n",
    "#     print(f\"Error: {str(e)}\\n\")\n",
    "#     raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b5ae50e6-4cd8-49d4-85d9-c9af4794ac80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Confirm that all sequence gaps have been removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e95e2dc9-1286-4d22-ad58-0268055d6f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add sequence motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02d02f2c-6d22-459c-9be1-9fa3e284a4ad",
   "metadata": {},
   "source": [
    "# Structure Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "742b2283-e178-4dc6-8c55-24e223461bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain and Store PDB files for non-gap sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "840173a6-6028-4bce-a4be-4e096bf2201b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obtain PDB files for gap sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "09a3b2c5-9bab-4f51-b529-f21c319ec920",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remodel using SCWRL4 and Store"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4c30448-49fe-4979-8fa7-68d73c26cc53",
   "metadata": {},
   "source": [
    "# Graph Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1023c659-8ca1-42bc-816c-6b46dd1579c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create NetworkX graphical representations of stored PDB files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ec4b8c55-3600-4443-9f7c-c7eedba7de72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split data into train/test splits based on homological similarity"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
